---
output: html_document
editor_options: 
  chunk_output_type: console
---
# Seurat Tutorials

Basic:
https://satijalab.org/seurat/articles/pbmc3k_tutorial.html
How to asign HTO info to seurat object:
https://satijalab.org/seurat/articles/multimodal_vignette.html
How to demultiplex data based on HTO:
https://satijalab.org/seurat/articles/hashing_vignette.html
To combine two 10x runs:
https://satijalab.org/seurat/articles/merge_vignette.html
To use SingleR:
https://www.singlecellcourse.org/single-cell-rna-seq-analysis-using-seurat.html#cell-type-annotation-using-singler
Pseudotime with monocle3:
https://rpubs.com/mahima_bose/Seurat_and_Monocle3_p


# Loading libraries

install.packages("Seurat")
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("SingleR")
BiocManager::install("celldex")
BiocManager::install("SingleCellExperiment")
BiocManager::install("DropletUtils")
BiocManager::install("dittoSeq")

```{r}
library(ggrepel)
library(Seurat)
library(ggplot2)
library(SingleR)
library(dplyr)
library(celldex)
library(RColorBrewer)
library(SingleCellExperiment)
library(DropletUtils)
library(Matrix)
library(knitr)
library(reshape2)
library(dittoSeq)
library(viridis)
library(RColorBrewer)
library(wesanderson)
library(DoubletFinder)
library(SeuratDisk)
library(httr)
library(scDblFinder)
library(tibble)
library(tidyr)
library(MAST)
library(patchwork)
library(ggsignif)
library(ggtext)
```

# Importing 10X data

The function Read10X() imports sparse matrix generated by Cellranger into the variable rawdata. The variable rawdata is a unique molecular identified (UMI) count matrix. The values in this matrix represent the number of molecules for each feature (i.e. gene; row) that are detected in each cell (column).

In our case, this object contain two list: 
- Gene Expression (GEX)
- Antibody capture (HTO)

```{r}
data_BM <- Read10X(data.dir = "../Data/sample_filtered_feature_bc_matrix_BM/")
```

```{r}
data_SP <- Read10X(data.dir = "../Data/sample_filtered_feature_bc_matrix_SP/")
```

# Create Seurat object

In the following code genes detected in less than 3 cells are filtered out.

```{r}
Seurat_Object_BM <- CreateSeuratObject(counts = data_BM$`Gene Expression`, project = "10x_A12_BM", min.cells = 3)
Seurat_Object_BM
```

```{r}
Seurat_Object_SP <- CreateSeuratObject(counts = data_SP$`Gene Expression`, project = "10x_A12_SP", min.cells = 3)
Seurat_Object_SP
```

## Add HTO information to seurat object

The seurat object contains an assay storing RNA measurement, but we add another assay with HTO info. 

```{r}
# create a new assay to store HTO information
HTO_assay <- CreateAssayObject(counts = data_BM$`Antibody Capture`)

# add this assay to the previously created Seurat object
Seurat_Object_BM$HTO <- HTO_assay

# Validate that the object now contains an assay with HTO info
rownames(Seurat_Object_BM$HTO)

# Note that all operations below are performed on the RNA assay Set and verify that the default assay is RNA.
DefaultAssay(Seurat_Object_BM)
```

```{r}
# create a new assay to store HTO information
HTO_assay <- CreateAssayObject(counts = data_SP$`Antibody Capture`)

# add this assay to the previously created Seurat object
Seurat_Object_SP$HTO <- HTO_assay

# Validate that the object now contains an assay with HTO info
rownames(Seurat_Object_SP$HTO)

# Note that all operations below are performed on the RNA assay Set and verify that the default assay is RNA.
DefaultAssay(Seurat_Object_SP)
```

### QUALITY CONTROL

First, we want to study the percentage of reads that map to mitochondrial genome, since Low-quality / dying cells often exhibit extensive mitochondrial contamination

```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Seurat_Object_BM[["percent.mt"]] <- PercentageFeatureSet(Seurat_Object_BM, pattern = "^mt-")
```

```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Seurat_Object_SP[["percent.mt"]] <- PercentageFeatureSet(Seurat_Object_SP, pattern = "^mt-")
```

## Visualize QC metrics:

```{r}
# violin plot
Idents(Seurat_Object_BM) <- Seurat_Object_BM$orig.ident
VlnPlot(Seurat_Object_BM, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(Seurat_Object_BM, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
mean(Seurat_Object_BM$nFeature_RNA)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Seurat_Object_BM, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Seurat_Object_BM, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

```{r}
# violin plot
Idents(Seurat_Object_SP) <- Seurat_Object_SP$orig.ident
VlnPlot(Seurat_Object_SP, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
mean(Seurat_Object_SP$nFeature_RNA)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot3 <- FeatureScatter(Seurat_Object_SP, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot4 <- FeatureScatter(Seurat_Object_SP, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot3 + plot4
```

## Other representations useful for quality control:

```{r}
counts_BM <- data_BM$`Gene Expression`
dim(counts)
counts_per_cell_BM <- Matrix::colSums(counts_BM)
counts_per_gene_BM <- Matrix::rowSums(counts_BM)
genes_per_cell_BM <- Matrix::colSums(counts_BM > 0) # count a gene only if it has non-zero reads mapped.
cells_per_gene_BM <- Matrix::rowSums(counts_BM > 0) # only count cells where the gene is expressed

hist(log10(counts_per_cell_BM + 1),main='counts per cell',col='blue')
hist(log10(genes_per_cell_BM + 1), main='genes per cell', col='blue')
plot(counts_per_cell_BM, genes_per_cell_BM, log='xy', col='blue') 
title('counts vs genes per cell')
hist(log10(counts_per_gene_BM + 1),main='counts per gene',col='blue')
```

```{r}
counts_SP<- data_SP$`Gene Expression`
dim(counts_SP)
counts_per_cell_SP <- Matrix::colSums(counts_SP)
counts_per_gene_SP <- Matrix::rowSums(counts_SP)
genes_per_cell_SP <- Matrix::colSums(counts_SP>0) # count a gene only if it has non-zero reads mapped.
cells_per_gene_SP <- Matrix::rowSums(counts_SP>0) # only count cells where the gene is expressed

hist(log10(counts_per_cell_SP+1),main='counts per cell',col='blue')
hist(log10(genes_per_cell_SP+1), main='genes per cell', col='blue')
plot(counts_per_cell_SP, genes_per_cell_SP, log='xy', col='blue') 
title('counts vs genes per cell')
hist(log10(counts_per_gene_SP+1),main='counts per gene',col='blue')
```

To establish the min and max genes per cells to avoid outliers: 
In our case we do not want to be very restrictive with max gene number since there are plasma cells with a lot of immunoglobuline transcripts.

```{r}
# Set upper and lower thresholds for genes per cell:
MIN_GENES_PER_CELL_BM <- 700  ## user-defined setting
MAX_GENES_PER_CELL_BM <- 6000  ## user-defined setting

# now replot with the thresholds being shown:
plot(sort(genes_per_cell_BM), xlab='cell', log='y', main='genes per cell (ordered)', 
  family = "Times New Roman")
abline(h=MIN_GENES_PER_CELL_BM, col='magenta')  # lower threshold
abline(h=MAX_GENES_PER_CELL_BM, col='gold') # upper threshold
```

```{r}
# Parameters
MIN_GENES_PER_CELL_SP <- 1000
MAX_GENES_PER_CELL_SP <- 6000

plot(
  sort(genes_per_cell_SP), 
  xlab = 'cell', 
  log = 'y', 
  main = 'genes per cell (ordered)', 
  family = "Times New Roman"  
)
abline(h = MIN_GENES_PER_CELL_SP, col = 'magenta')  # lower threshold
abline(h = MAX_GENES_PER_CELL_SP, col = 'gold')     # upper threshold
```

To establish the threshold for mithocondrial genes:

```{r}
# define the mitochondrial genes
mito_genes_BM <- grep("^mt-", rownames(counts_BM) , ignore.case=T, value=T)
print(mito_genes_BM)
# compute pct mito
mito_gene_read_counts_BM = Matrix::colSums(counts_BM[mito_genes_BM,])
pct_mito_BM = mito_gene_read_counts_BM / counts_per_cell_BM * 100

MAX_PCT_MITO_BM <- 5   ## user-defined setting

plot(sort(pct_mito_BM), xlab = "cells sorted by percentage mitochondrial counts", ylab = "percentage mitochondrial counts")
abline(h=MAX_PCT_MITO_BM, col='red')
```

```{r}
# define the mitochondrial genes
mito_genes_SP <- grep("^mt-", rownames(counts_SP) , ignore.case=T, value=T)
print(mito_genes_SP)
# compute pct mito
mito_gene_read_counts_SP = Matrix::colSums(counts_SP[mito_genes_SP,])
pct_mito_SP = mito_gene_read_counts_SP / counts_per_cell_SP * 100

MAX_PCT_MITO_SP <- 4.5   ## user-defined setting

plot(sort(pct_mito_SP), xlab = "cells sorted by percentage mitochondrial counts", ylab = "percentage mitochondrial counts")
abline(h=MAX_PCT_MITO_SP, col='red')
```

## Filtering based on QC metrics

Filtering is a subjective step which totally depends on the type of data one has.
Based on the previous plots we have established the following filters:
We filter cells that have unique feature counts over 6000 or less than 500
We filter cells that have >5% mitochondrial counts

```{r}
Seurat_Object_BM <- subset(Seurat_Object_BM, subset = nFeature_RNA > 700 & nFeature_RNA < 6000 & percent.mt < 5)
```

```{r}
Seurat_Object_SP <- subset(Seurat_Object_SP, subset = nFeature_RNA > 1000 & nFeature_RNA < 6000 & percent.mt < 4.5)
```

# Normalizing the data

By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. Normalized values are stored in pbmc[["RNA"]]@data.

LOGNORMALIZE Method Steps:

- Library Size Factor Calculation: For each cell, the total expression (sum of counts) is calculated. This total is used as a library size factor.

- Scaling: Each gene count in a cell is divided by the total counts (library size) of that cell. This step converts the counts into relative proportions (or fractions) of the total counts for each cell.

- Scaling to a Common Value: These proportions are then multiplied by a scale factor (often set to 10,000) to bring them to a common scale. This helps in making the data more interpretable and comparable across cells.

- Log Transformation: Finally, a logarithmic transformation is applied, typically log1p, which is log(1 + x). This transformation helps in compressing the range of the data and stabilizing the variance.

On the other hand, HTO or ADT data is normalized with the "CLR" method:

- Very useful for Compositional Data
Constant Sum: The sum of all parts (e.g., all gene counts in a cell) is constant. This sum can vary between samples, but within a sample, the sum of all parts is fixed.
Relative Proportions: The important aspect is not the absolute value of each part, but its relative proportion compared to the total.
Interdependence of Parts: Since the sum of all parts is constant, an increase in one part must be compensated by a decrease in another part(s).

We normalize both GEX and HTO data

GEX is normalized by "LogNormalize" (default)

```{r}
Seurat_Object_BM <- NormalizeData(Seurat_Object_BM)
Seurat_Object_BM <- NormalizeData(Seurat_Object_BM, assay = "HTO", normalization.method = "CLR")
```

```{r}
Seurat_Object_SP <- NormalizeData(Seurat_Object_SP)
Seurat_Object_SP <- NormalizeData(Seurat_Object_SP, assay = "HTO", normalization.method = "CLR")
```

# Demultiplex data based on HTO enrichment

Demultiplex cells and identify cross-sample doublets.
Observe the number of singlets, doublets and negative cells
https://satijalab.org/seurat/articles/multimodal_vignette

The number written in HTODemux (0.99) has to be carefully selected. If this number is very high, only cells with very high reads of HTO will be considered as positive. This increases doublets (as it is difficult to be considered a positive cell, cells with WT-3 high and WT-4 very-high for example, will only be considered as WT4). Still, we need to find an equilibrium being restrictive but not in excess.

I have chosen to set this threshold to 0.9999 in spleen as the distribution of HTO counts is more clear and we need to be restrictive in taking cells that are real huIgk+. In bone marrow, we have been less restrictive due to the fact that hto counts for some mice were low in some cases.

```{r}
Seurat_Object_BM <- HTODemux(Seurat_Object_BM, assay = "HTO", positive.quantile = 0.99)
```

```{r}
Seurat_Object_SP <- HTODemux(Seurat_Object_SP, assay = "HTO", positive.quantile = 0.99999)

# Global classification results
table_result <- table(Seurat_Object_SP$HTO_classification)
table_result_sorted <- sort(table_result, decreasing = FALSE)
table_result_sorted
```

Verify that the positive quantile is selected properly

```{r}
plot1 <- FeatureScatter(Seurat_Object_SP, feature1 = "hto_SP-A12-1", feature2 = "hto_huK", pt.size = 1)
plot2 <- FeatureScatter(Seurat_Object_SP, feature1 = "hto_SP-A12-2", feature2 = "hto_huK", pt.size = 1)
plot3 <- FeatureScatter(Seurat_Object_SP, feature1 = "hto_SP-A12-3", feature2 = "hto_huK", pt.size = 1)
plot4 <- FeatureScatter(Seurat_Object_SP, feature1 = "hto_SP-WT-1", feature2 = "hto_huK", pt.size = 1)
plot5 <- FeatureScatter(Seurat_Object_SP, feature1 = "hto_SP-WT-2", feature2 = "hto_huK", pt.size = 1)
plot6 <- FeatureScatter(Seurat_Object_SP, feature1 = "hto_SP-WT-3", feature2 = "hto_huK", pt.size = 1)
(plot1 + plot2 + plot3 + plot4 + plot5 + plot6) 
```

```{r}
plot1 <- FeatureScatter(Seurat_Object_BM, feature1 = "hto_BM-A12-1", feature2 = "hto_huK", pt.size = 1)
plot2 <- FeatureScatter(Seurat_Object_BM, feature1 = "hto_BM-A12-2", feature2 = "hto_huK", pt.size = 1)
plot3 <- FeatureScatter(Seurat_Object_BM, feature1 = "hto_BM-A12-3", feature2 = "hto_huK", pt.size = 1)
plot4 <- FeatureScatter(Seurat_Object_BM, feature1 = "hto_BM-WT-1", feature2 = "hto_huK", pt.size = 1)
plot5 <- FeatureScatter(Seurat_Object_BM, feature1 = "hto_BM-WT-2", feature2 = "hto_huK", pt.size = 1)
plot6 <- FeatureScatter(Seurat_Object_BM, feature1 = "hto_BM-WT-3", feature2 = "hto_huK", pt.size = 1)
(plot1 + plot2 + plot3 + plot4 + plot5 + plot6) 
```

Let’s convert our Seurat object to single cell experiment (SCE) for convenience.
```{r}
sce_SP <- as.SingleCellExperiment(DietSeurat(Seurat_Object_SP))
sce_SP

sce_BM <- as.SingleCellExperiment(DietSeurat(Seurat_Object_BM))
sce_BM
```



# scDblFinder

Doublet depletion
scDblFinder has two main modes for generating artificial doublets: a random one (clusters=FALSE, now default) and a cluster-based one (clusters=TRUE or providing your own clusters - the approach from previous versions). In practice, we observed that both approaches perform well (and better than alternatives). We suggest using the cluster-based approach when the datasets are segregated into clear clusters, and the random one for the rest (e.g. developmental trajectories).

```{r}
sce_BM <- scDblFinder(sce_BM)
table(sce_BM$scDblFinder.class)
doublets_scDblFinder_BM <- colnames(sce_BM[, sce_BM$scDblFinder.class == "doublet"])

sce_SP <- scDblFinder(sce_SP)
table(sce_SP$scDblFinder.class)
doublets_scDblFinder_SP <- colnames(sce_SP[, sce_SP$scDblFinder.class == "doublet"])
```

# Eliminating doublets from HTO analysis with scDblFinder

```{r}
Seurat_Object_BM_singlets <- subset(Seurat_Object_BM, cells = setdiff(Cells(Seurat_Object_BM), doublets_scDblFinder_BM))
Seurat_Object_SP_singlets <- subset(Seurat_Object_SP, cells = setdiff(Cells(Seurat_Object_SP), doublets_scDblFinder_SP))
```

# HTO signal

Group cells based on the max HTO signal

```{r}
Idents(Seurat_Object_BM) <- "HTO_maxID"
RidgePlot(Seurat_Object_BM, assay = "HTO", features = rownames(Seurat_Object_BM[["HTO"]])[1:7], ncol = 2)
```

```{r}
Idents(Seurat_Object_SP) <- "HTO_maxID"
RidgePlot(Seurat_Object_SP, assay = "HTO", features = rownames(Seurat_Object_SP[["HTO"]])[1:7], ncol = 2)
```



Compare number of UMIs for singlets, doublets and negative cells
```{r}
Idents(Seurat_Object_BM) <- "HTO_classification"
VlnPlot(Seurat_Object_BM, features = "nCount_RNA", pt.size = 0.1, log = TRUE)
VlnPlot(Seurat_Object_BM_singlets, features = "nCount_RNA", pt.size = 0.1, log = TRUE)
```

```{r}
Idents(Seurat_Object_SP) <- "HTO_classification"
VlnPlot(Seurat_Object_SP, features = "nCount_RNA", pt.size = 0.1, log = TRUE)
VlnPlot(Seurat_Object_SP_singlets, features = "nCount_RNA", pt.size = 0.1, log = TRUE)
```

Compare number of UMIs for combinations of different antibodies
```{r}
my_palette <- viridis(29)
VlnPlot(Seurat_Object_BM, features = "nCount_RNA", pt.size = 0.1, log = TRUE) +
  scale_fill_manual(values = my_palette)
VlnPlot(Seurat_Object_BM_singlets, features = "nCount_RNA", pt.size = 0.1, log = TRUE) +
  scale_fill_manual(values = my_palette)
```


```{r}
Idents(Seurat_Object_SP) <- "HTO_classification"
Idents(Seurat_Object_SP_singlets) <- "HTO_classification"
VlnPlot(Seurat_Object_SP, features = "nCount_RNA", pt.size = 0.1, log = TRUE)+
  scale_fill_manual(values = my_palette)
VlnPlot(Seurat_Object_SP_singlets, features = "nCount_RNA", pt.size = 0.1, log = TRUE)+
  scale_fill_manual(values = my_palette)
```

# Keep HTO combinations that reflect singlets

```{r}
Seurat_Object_BM_selected <- subset(Seurat_Object_BM_singlets, HTO_classification %in% c("BM-WT-2", "BM-WT-3", "BM-A12-3", "BM-A12-1", "BM-A12-2", "BM-WT-1", "BM-A12-1_huK", "BM-A12-3_huK", "BM-A12-1_huK", "BM-WT-2_huK", "BM-A12-2_huK", "BM-WT-1_huK", "BM-WT-3_huK"))
Seurat_Object_BM_selected 
Idents(Seurat_Object_BM_selected) <- "HTO_classification"
RidgePlot(Seurat_Object_BM_selected, assay = "HTO", features = rownames(Seurat_Object_BM_selected[["HTO"]])[1:6], ncol = 2)
```

```{r}
Seurat_Object_SP_selected <- subset(Seurat_Object_SP_singlets, HTO_classification %in% c("SP-WT-3", "SP-WT-2", "SP-A12-3", "SP-A12-2", "SP-WT-1", "SP-A12-1", "huK_SP-A12-1", "huK_SP-A12-2", "huK_SP-A12-3", "huK_SP-WT-3", "huK_SP-WT-2", "huK_SP-WT-1"))
Seurat_Object_SP_selected 
Idents(Seurat_Object_SP_selected) <- "HTO_classification"
RidgePlot(Seurat_Object_SP_selected, assay = "HTO", features = rownames(Seurat_Object_SP_selected[["HTO"]])[1:6], ncol = 2)
```

We are going to verify if doublets make sense (combination of huIgk with A12 HTOs)

```{r}
plot1 <- FeatureScatter(Seurat_Object_SP_selected, feature1 = "hto_SP-A12-1", feature2 = "hto_huK", pt.size = 1)
plot2 <- FeatureScatter(Seurat_Object_SP_selected, feature1 = "hto_SP-A12-2", feature2 = "hto_huK", pt.size = 1)
plot3 <- FeatureScatter(Seurat_Object_SP_selected, feature1 = "hto_SP-A12-3", feature2 = "hto_huK", pt.size = 1)
plot4 <- FeatureScatter(Seurat_Object_SP_selected, feature1 = "hto_SP-WT-1", feature2 = "hto_huK", pt.size = 1)
plot5 <- FeatureScatter(Seurat_Object_SP_selected, feature1 = "hto_SP-WT-2", feature2 = "hto_huK", pt.size = 1)
plot6 <- FeatureScatter(Seurat_Object_SP_selected, feature1 = "hto_SP-WT-3", feature2 = "hto_huK", pt.size = 1)
(plot1 + plot2 + plot3 + plot4 + plot5 + plot6) 
```

```{r}
plot1 <- FeatureScatter(Seurat_Object_BM_selected, feature1 = "hto_BM-A12-1", feature2 = "hto_huK", pt.size = 1)
plot2 <- FeatureScatter(Seurat_Object_BM_selected, feature1 = "hto_BM-A12-2", feature2 = "hto_huK", pt.size = 1)
plot3 <- FeatureScatter(Seurat_Object_BM_selected, feature1 = "hto_BM-A12-3", feature2 = "hto_huK", pt.size = 1)
plot4 <- FeatureScatter(Seurat_Object_BM_selected, feature1 = "hto_BM-WT-1", feature2 = "hto_huK", pt.size = 1)
plot5 <- FeatureScatter(Seurat_Object_BM_selected, feature1 = "hto_BM-WT-2", feature2 = "hto_huK", pt.size = 1)
plot6 <- FeatureScatter(Seurat_Object_BM_selected, feature1 = "hto_BM-WT-3", feature2 = "hto_huK", pt.size = 1)
(plot1 + plot2 + plot3 + plot4 + plot5 + plot6) 
```

# Identification of highly variable features (feature selection)

We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). Previous analysis have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.

By default FindVariableFeatures() function return 2,000 features per dataset. These will be used in downstream analysis, like PCA.

```{r}
DefaultAssay(Seurat_Object_SP_selected) <- "RNA"
Seurat_Object_SP_selected  <- FindVariableFeatures(Seurat_Object_SP_selected , selection.method = "vst", nfeatures = 3000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(Seurat_Object_SP_selected), 10)

# plot variable features with labels
plot1 <- VariableFeaturePlot(Seurat_Object_SP_selected)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```


```{r}
DefaultAssay(Seurat_Object_BM_selected) <- "RNA"
Seurat_Object_BM_selected <- FindVariableFeatures(Seurat_Object_BM_selected, selection.method = "vst", nfeatures = 3000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(Seurat_Object_BM_selected), 10)

# plot variable features with labels
plot1 <- VariableFeaturePlot(Seurat_Object_BM_selected)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```


# Scaling the data

Next, we apply a linear transformation (‘scaling’) that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The ScaleData() function:
- Shifts the expression of each gene, so that the mean expression across cells is 0
- Scales the expression of each gene, so that the variance across cells is 1 (This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate)
- The results of this are stored in pbmc[["RNA"]]@scale.data

Although for PCA and clustering we are going to use only the Variable features, for other analysis such as Seurat Heatmaps it is important to have all the genes in the heatmeap to be scaled so the highly-expressed genes don't dominate the heatmap. That's hay we normalize all the genes. 

NormalizeData

The NormalizeData function is used to normalize gene expression data within each cell. Normalizing the data is crucial in scRNA-seq to ensure that differences in sequencing depth (number of reads per cell) and cell size do not improperly affect the analysis results. The main approaches to normalization include:

Total Normalization: Divides the gene expression of each cell by the total gene expression for that cell, thereby normalizing for sequencing depth.

Reference Cell Normalization: Divides the gene expression of each cell by that of a reference cell (e.g., a cell with average gene expression).

ScaleData

The ScaleData function (also known as standardization) is applied after normalization to transform and standardize gene expression data to a common scale. This ensures that all features (genes) have comparable variance before performing statistical analyses like PCA (Principal Component Analysis). Standardization typically involves:

Z-Score Standardization: Subtracts the mean and divides by the standard deviation for each gene across all cells, so that each gene has a mean of zero and a standard deviation of one.


```{r}
Seurat_Object_SP_selected <- ScaleData(Seurat_Object_SP_selected, features = VariableFeatures(Seurat_Object_BM_selected))
```

```{r}
Seurat_Object_BM_selected <- ScaleData(Seurat_Object_BM_selected, features = VariableFeatures(Seurat_Object_BM_selected))
```


# PCA

Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using features argument if you wish to choose a different subset.
We deplete the A12 gene from the clustering analysis in order to avoid bias.

```{r}
variable_genes <- setdiff(VariableFeatures(Seurat_Object_SP_selected), c("A12"))
Seurat_Object_SP_selected <- RunPCA(Seurat_Object_SP_selected, features = variable_genes)
```

```{r}
variable_genes <- setdiff(VariableFeatures(Seurat_Object_BM_selected), c("A12"))
Seurat_Object_BM_selected <- RunPCA(Seurat_Object_BM_selected, features = variable_genes)
```

# Cluster the cells

we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors() function, and takes as input the previously defined dimensionality of the dataset (first 18 PCs).

To cluster the cells, we next apply modularity optimization techniques to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters() function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. Setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the Idents() function.


## Cell Type annotation using SingleR

We used the ImmGen database in order to automatically annotate immune cells in our data.

```{r}
# Download Immgen data
immgen.se <-ImmGenData()
```

```{r}
sce_SP <- as.SingleCellExperiment(DietSeurat(Seurat_Object_SP_selected))
sce_SP

sce_BM <- as.SingleCellExperiment(DietSeurat(Seurat_Object_BM_selected))
sce_BM
```

### Label main
To study the main immune cell types

SingleR es una herramienta bioinformática utilizada para realizar análisis de transferencia de conocimiento en datos de expresión génica, especialmente en el contexto de datos de secuenciación de ARN (RNA-seq) de célula única. Su objetivo principal es comparar perfiles de expresión génica de un conjunto de datos de prueba con perfiles de expresión de referencia de conjuntos de datos públicos o previamente caracterizados.
```{r}
immgen_SP.main <- SingleR(test = sce_SP,assay.type.test = 1,ref = immgen.se,labels = immgen.se$label.main)
table(immgen_SP.main$pruned.labels)
```

```{r}
immgen_BM.main <- SingleR(test = sce_BM,assay.type.test = 1,ref = immgen.se,labels = immgen.se$label.main)
table(immgen_BM.main$pruned.labels)
```

### Add annotations to seurat object

```{r}
Seurat_Object_SP_selected@meta.data$immgen_SP.main <- immgen_SP.main$pruned.labels
```

```{r}
Seurat_Object_BM_selected@meta.data$immgen_BM.main <- immgen_BM.main$pruned.labels
```

See where are the non-B cells in the umap plot:
```{r}
DimPlot(Seurat_Object_SP_selected, reduction = "pca", group.by = "immgen_SP.main")
table(Seurat_Object_SP_selected$immgen_SP.main)
```

```{r}
DimPlot(Seurat_Object_BM_selected, reduction = "pca", group.by = "immgen_BM.main")
table(Seurat_Object_BM_selected$immgen_BM.main)
```

Calculate the number of cells per cell type

```{r}

# Calculate the number of cells per group
cell_counts <- table(Seurat_Object_BM_selected$immgen_BM.main)

# Create labels by combining the group name and the number of cells
new_labels <- paste0(names(cell_counts), " (", cell_counts, ")")
names(new_labels) <- names(cell_counts)  # Assign the original names as names for the new labels

# Create the DimPlot with automatic colors, title, and customized legend
DimPlot(
    Seurat_Object_BM_selected, 
    reduction = "pca", 
    group.by = "immgen_BM.main"
) +
    scale_color_discrete(labels = new_labels) +  # Labels with cell counts and automatic colors
    ggtitle("PCA reduction with immune cell annotations from Immgen") +  # Title
    theme(
        legend.text = element_text(size = 10, family = "Times New Roman"),  # Legend font
        legend.title = element_text(size = 12, family = "Times New Roman"), # Legend title font
        plot.title = element_text(size = 14, family = "Times New Roman")    # Title customization
    )
```

### Asign cell type to cell identity

```{r}
Idents(Seurat_Object_SP_selected) <- Seurat_Object_SP_selected$immgen_SP.main
```

```{r}
Idents(Seurat_Object_BM_selected) <- Seurat_Object_BM_selected$immgen_BM.main
```

## Select only B cells

```{r}
Seurat_Object_SP_selected_Bcells<- subset(Seurat_Object_SP_selected, idents = c("B cells"))
Seurat_Object_SP_selected_Bcells
```

```{r}
Seurat_Object_BM_selected_Bcells<- subset(Seurat_Object_BM_selected, idents = c("B cells", "B cells, pro"))
Seurat_Object_BM_selected_Bcells
```



# Redo Scaling, PCA and clustering for B cells

## Identifucation of highly variable features of B cells

```{r}
Seurat_Object_SP_selected_Bcells <- FindVariableFeatures(Seurat_Object_SP_selected_Bcells, selection.method = "vst", nfeatures = 3000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(Seurat_Object_SP_selected_Bcells), 10)

# plot variable features with labels
plot1 <- VariableFeaturePlot(Seurat_Object_SP_selected_Bcells)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```

```{r}
Seurat_Object_BM_selected_Bcells <- FindVariableFeatures(Seurat_Object_BM_selected_Bcells, selection.method = "vst", nfeatures = 3000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(Seurat_Object_BM_selected_Bcells), 10)

# plot variable features with labels
plot1 <- VariableFeaturePlot(Seurat_Object_BM_selected_Bcells)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```
## Scale B cells

```{r}
variable_genes_SP <- setdiff(VariableFeatures(Seurat_Object_SP_selected_Bcells), c("A12"))
Seurat_Object_SP_selected_Bcells <- ScaleData(Seurat_Object_SP_selected_Bcells, features = row.names(Seurat_Object_SP_selected_Bcells)) 
```

```{r}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

Seurat_Object_BM_selected_Bcells <- CellCycleScoring(Seurat_Object_BM_selected_Bcells, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)


Seurat_Object_BM_selected_Bcells$CC.Difference <- Seurat_Object_BM_selected_Bcells$S.Score - Seurat_Object_BM_selected_Bcells$G2M.Score


variable_genes_BM <- setdiff(VariableFeatures(Seurat_Object_BM_selected_Bcells), c("A12"))
Seurat_Object_BM_selected_Bcells <- ScaleData(Seurat_Object_BM_selected_Bcells, vars.to.regress = c("CC.Difference"), features = row.names(Seurat_Object_BM_selected_Bcells)) 
```

## PCA of B cells

```{r}
Seurat_Object_SP_selected_Bcells <- RunPCA(Seurat_Object_SP_selected_Bcells, features = variable_genes_SP) #
```

```{r}
Seurat_Object_BM_selected_Bcells <- RunPCA(Seurat_Object_BM_selected_Bcells, features = variable_genes_BM) 
```

Representation:
```{r}
DimPlot(Seurat_Object_SP_selected_Bcells, reduction = "pca")
```

```{r}
DimPlot(Seurat_Object_BM_selected_Bcells, reduction = "pca")
```

### Determine the ‘dimensionality’ of the dataset and obtain evidence to choose number of principal components

```{r}
Seurat_Object_SP_selected_Bcells <- JackStraw(Seurat_Object_SP_selected_Bcells, num.replicate = 100, dims = 40)
```

```{r}
Seurat_Object_BM_selected_Bcells <- JackStraw(Seurat_Object_BM_selected_Bcells, num.replicate = 100, dims = 40)
```

```{r}
Seurat_Object_SP_selected_Bcells <- ScoreJackStraw(Seurat_Object_SP_selected_Bcells, dims = 1:40)
```

```{r}
Seurat_Object_BM_selected_Bcells <- ScoreJackStraw(Seurat_Object_BM_selected_Bcells, dims = 1:40)
```

```{r}
JackStrawPlot(Seurat_Object_SP_selected_Bcells, dims = 1:40)
```

```{r}
JackStrawPlot(Seurat_Object_BM_selected_Bcells, dims = 1:40) # No has corrido esto
```

An heuristic method generates an ‘Elbow plot’: a ranking of principle components based on the percentage of variance explained by each one (ElbowPlot() function). 

```{r}
ElbowPlot(Seurat_Object_SP_selected_Bcells, ndims=40)
```

```{r}
ElbowPlot(Seurat_Object_BM_selected_Bcells, ndims=40)
```

# Cluster the cells

we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors() function, and takes as input the previously defined dimensionality of the dataset.

To cluster the cells, we next apply modularity optimization techniques to iteratively group cells together, with the goal of optimizing the standard modularity function. The FindClusters() function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. 

## B cells clusters

```{r}
Seurat_Object_SP_selected_Bcells <- FindNeighbors(Seurat_Object_SP_selected_Bcells, dims = 1:20)
Seurat_Object_SP_selected_Bcells <- FindClusters(Seurat_Object_SP_selected_Bcells, resolution = 0.6)
```

## B cells clusters

```{r}
Seurat_Object_BM_selected_Bcells <- FindNeighbors(Seurat_Object_BM_selected_Bcells, dims = 1:15)
Seurat_Object_BM_selected_Bcells <- FindClusters(Seurat_Object_BM_selected_Bcells, resolution = 0.9)
```

Run non-linear dimensional reduction (UMAP/tSNE)

```{r}
Seurat_Object_SP_selected_Bcells <- RunUMAP(Seurat_Object_SP_selected_Bcells, dims = 1:20)
DimPlot(Seurat_Object_SP_selected_Bcells, reduction = "umap", label=T, group.by = "seurat_clusters")
```

```{r}
Seurat_Object_BM_selected_Bcells <- RunUMAP(Seurat_Object_BM_selected_Bcells, dims = 1:15)
DimPlot(Seurat_Object_BM_selected_Bcells, reduction = "umap", label=T, group.by = "seurat_clusters")
```


# Assign the corresponding genotype to each cell

```{r}
Seurat_Object_SP_selected_Bcells$genotype <- "NA"
 
A12_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("SP-A12-1","SP-A12-2","SP-A12-3","huK_SP-A12-2","huK_SP-A12-1","huK_SP-A12-3"))]
Seurat_Object_SP_selected_Bcells@meta.data[A12_cells, "genotype"] = "A12"

WT_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("SP-WT-1","SP-WT-2","SP-WT-3","huK_SP-WT-2","huK_SP-WT-1","huK_SP-WT-3"))]
Seurat_Object_SP_selected_Bcells@meta.data[WT_cells, "genotype"] = "WT"
```

```{r}
DimPlot(Seurat_Object_SP_selected_Bcells, reduction = "umap", split.by = "genotype", group.by = "seurat_clusters")
```

```{r}
Seurat_Object_BM_selected_Bcells$genotype <- "NA"
 
A12_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-A12-1","BM-A12-2","BM-A12-3","BM-A12-2_huK","BM-A12-1_huK","BM-A12-3_huK"))]
Seurat_Object_BM_selected_Bcells@meta.data[A12_cells, "genotype"] = "A12"

WT_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-WT-1","BM-WT-2","BM-WT-3","BM-WT-1_huK","BM-WT-2_huK","BM-WT-3_huK"))]
Seurat_Object_BM_selected_Bcells@meta.data[WT_cells, "genotype"] = "WT"
```

```{r}
Seurat_Object_BM_selected_Bcells$genotype <- factor(Seurat_Object_BM_selected_Bcells$genotype, levels = c("WT", "A12"))
Seurat_Object_SP_selected_Bcells$genotype <- factor(Seurat_Object_SP_selected_Bcells$genotype, levels = c("WT", "A12"))
```


```{r}
Idents(Seurat_Object_BM_selected_Bcells) <- Seurat_Object_BM_selected_Bcells$seurat_clusters
DimPlot(Seurat_Object_BM_selected_Bcells, reduction = "umap", split.by = "genotype")
```

# Assign huIgk CITE-seq information to each cell

```{r}
Seurat_Object_SP_selected_Bcells$huIgk <- "NA"
 
huk_A12_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("huK_SP-A12-2","huK_SP-A12-1","huK_SP-A12-3"))]
Seurat_Object_SP_selected_Bcells@meta.data[huk_A12_cells, "huIgk"] = "huk_A12"

huk_WT_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("huK_SP-WT-2","huK_SP-WT-1","huK_SP-WT-3"))]
Seurat_Object_SP_selected_Bcells@meta.data[huk_WT_cells, "huIgk"] = "huk_WT"

neg_A12_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("SP-A12-1","SP-A12-2","SP-A12-3"))]
Seurat_Object_SP_selected_Bcells@meta.data[neg_A12_cells, "huIgk"] = "neg_A12"

neg_WT_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("SP-WT-1","SP-WT-2","SP-WT-3"))]
Seurat_Object_SP_selected_Bcells@meta.data[neg_WT_cells, "huIgk"] = "neg_WT"

table(Seurat_Object_SP_selected_Bcells$huIgk)
```

```{r}
DimPlot(Seurat_Object_SP_selected_Bcells, reduction = "umap", split.by = "huIgk")

```

```{r}
Seurat_Object_BM_selected_Bcells$huIgk <- "NA"
 
huk_A12_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-A12-2_huK","BM-A12-1_huK","BM-A12-3_huK"))]
Seurat_Object_BM_selected_Bcells@meta.data[huk_A12_cells, "huIgk"] = "huk_A12"

huk_WT_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-WT-1_huK","BM-WT-2_huK","BM-WT-3_huK"))]
Seurat_Object_BM_selected_Bcells@meta.data[huk_WT_cells, "huIgk"] = "huk_WT"

neg_A12_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-A12-1","BM-A12-2","BM-A12-3"))]
Seurat_Object_BM_selected_Bcells@meta.data[neg_A12_cells, "huIgk"] = "neg_A12"

neg_WT_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-WT-1","BM-WT-2","BM-WT-3"))]
Seurat_Object_BM_selected_Bcells@meta.data[neg_WT_cells, "huIgk"] = "neg_WT"

table(Seurat_Object_BM_selected_Bcells$huIgk)
```

```{r}
DimPlot(Seurat_Object_BM_selected_Bcells, reduction = "umap", split.by = "huIgk")

```

```{r}
Seurat_Object_BM_selected_Bcells$huIgk <- factor(Seurat_Object_BM_selected_Bcells$huIgk, levels = c("neg_WT", "huk_WT", "neg_A12", "huk_A12"))
Seurat_Object_SP_selected_Bcells$huIgk <- factor(Seurat_Object_SP_selected_Bcells$huIgk, levels = c("neg_WT", "huk_WT", "neg_A12", "huk_A12"))
```

# Assign the corresponding mice of origin to each cell

```{r}
Seurat_Object_SP_selected_Bcells$mice <- "NA"
 
WT_1_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("huK_SP-WT-1","SP-WT-1"))]
Seurat_Object_SP_selected_Bcells@meta.data[WT_1_cells, "mice"] = "WT_1"

WT_2_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("huK_SP-WT-2","SP-WT-2"))]
Seurat_Object_SP_selected_Bcells@meta.data[WT_2_cells, "mice"] = "WT_2"

WT_3_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("huK_SP-WT-3","SP-WT-3"))]
Seurat_Object_SP_selected_Bcells@meta.data[WT_3_cells, "mice"] = "WT_3"

A12_1_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("huK_SP-A12-1","SP-A12-1"))]
Seurat_Object_SP_selected_Bcells@meta.data[A12_1_cells, "mice"] = "A12_1"

A12_2_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("huK_SP-A12-2","SP-A12-2"))]
Seurat_Object_SP_selected_Bcells@meta.data[A12_2_cells, "mice"] = "A12_2"

A12_3_cells <- Cells(Seurat_Object_SP_selected_Bcells)[which(Seurat_Object_SP_selected_Bcells$HTO_classification %in% c("huK_SP-A12-3","SP-A12-3"))]
Seurat_Object_SP_selected_Bcells@meta.data[A12_3_cells, "mice"] = "A12_3"

table(Seurat_Object_SP_selected_Bcells$mice)
```

```{r}
DimPlot(Seurat_Object_SP_selected_Bcells, reduction = "umap", split.by = "mice")
```

```{r}
Seurat_Object_BM_selected_Bcells$mice <- "NA"
 
WT_1_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-WT-1_huK","BM-WT-1"))]
Seurat_Object_BM_selected_Bcells@meta.data[WT_1_cells, "mice"] = "WT_1"

WT_2_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-WT-2_huK","BM-WT-2"))]
Seurat_Object_BM_selected_Bcells@meta.data[WT_2_cells, "mice"] = "WT_2"

WT_3_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-WT-3_huK","BM-WT-3"))]
Seurat_Object_BM_selected_Bcells@meta.data[WT_3_cells, "mice"] = "WT_3"

A12_1_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-A12-1_huK","BM-A12-1"))]
Seurat_Object_BM_selected_Bcells@meta.data[A12_1_cells, "mice"] = "A12_1"

A12_2_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-A12-2_huK","BM-A12-2"))]
Seurat_Object_BM_selected_Bcells@meta.data[A12_2_cells, "mice"] = "A12_2"

A12_3_cells <- Cells(Seurat_Object_BM_selected_Bcells)[which(Seurat_Object_BM_selected_Bcells$HTO_classification %in% c("BM-A12-3_huK","BM-A12-3"))]
Seurat_Object_BM_selected_Bcells@meta.data[A12_3_cells, "mice"] = "A12_3"

table(Seurat_Object_BM_selected_Bcells$mice)
```

```{r}
DimPlot(Seurat_Object_BM_selected_Bcells, reduction = "umap", split.by = "mice")
```

```{r}
Seurat_Object_BM_selected_Bcells$mice <- factor(Seurat_Object_BM_selected_Bcells$mice, levels = c("WT_1", "WT_2", "WT_3", "A12_1","A12_2","A12_3"))
Seurat_Object_SP_selected_Bcells$mice <- factor(Seurat_Object_SP_selected_Bcells$mice, levels = c("WT_1", "WT_2", "WT_3", "A12_1","A12_2","A12_3"))
```

### Immgen fine A12

Focus on specific B cell subsets present in each organ (spleen or bone marrow)

Let’s convert our Seurat object to single cell experiment (SCE) for convenience.
```{r}
sce_SP_Bcells <- as.SingleCellExperiment(DietSeurat(Seurat_Object_SP_selected_Bcells))
sce_SP_Bcells
```

```{r}
sce_BM_Bcells <- as.SingleCellExperiment(DietSeurat(Seurat_Object_BM_selected_Bcells))
sce_BM_Bcells
```

```{r}
bcell_filter <- immgen.se$label.main %in% c("B cells", "B cells, pro")
immgen_Bcells.se <- immgen.se[, bcell_filter]

# Only bone marrow B cell populations
bone_marrow_filter <- immgen_Bcells.se$label.fine %in% c("B cells (B.FrE)", "B cells (preB.FrD)", "B cells (B.FrF)", "B cells (preB.FrC)", "B cells (proB.FrBC)", "B cells, pro (proB.FrA)", "B cells (proB.CLP)")
immgen_Bcells_BoneMarrow.se <- immgen_Bcells.se[, bone_marrow_filter]

# Only splenic B cell populations
SP_filter <- immgen_Bcells.se$label.fine %in% c("B cells (B.Fo)", "B cells (B.GC)","B cells (B.MZ)", "B cells (B.MEM)", "B cells (B.T1)", "B cells (B.T2)", "B cells (B.T3)", "B cells (B1b)", "B cells (B1a)")
immgen_Bcells_SP.se <- immgen_Bcells.se[, SP_filter]
```


```{r}
immgen_SP.fine <- SingleR(test = sce_SP_Bcells, assay.type.test = 1,ref = immgen_Bcells_SP.se ,labels = immgen_Bcells_SP.se$label.fine)
table(immgen_SP.fine$pruned.labels)
```

```{r}
immgen_BM.fine <- SingleR(test = sce_BM_Bcells, assay.type.test = 1,ref = immgen_Bcells_BoneMarrow.se,labels = immgen_Bcells_BoneMarrow.se$label.fine)
table(immgen_BM.fine$pruned.labels)
```
### Add annotations to seurat object

```{r}
Seurat_Object_SP_selected_Bcells@meta.data$immgen_SP.fine <- immgen_SP.fine$pruned.labels
Seurat_Object_SP_selected_Bcells$immgen_SP.fine <- na.omit(Seurat_Object_SP_selected_Bcells$immgen_SP.fine)
```

```{r}
Seurat_Object_BM_selected_Bcells@meta.data$immgen_BM.fine <- immgen_BM.fine$pruned.labels
Seurat_Object_BM_selected_Bcells$immgen_BM.fine <- na.omit(Seurat_Object_BM_selected_Bcells$immgen_BM.fine)
```

```{r}
DimPlot(Seurat_Object_SP_selected_Bcells, reduction = "umap", group.by = "immgen_SP.fine", split.by = "genotype")

DimPlot(Seurat_Object_BM_selected_Bcells, reduction = "umap", group.by = "immgen_BM.fine", split.by = "genotype")

DimPlot(Seurat_Object_SP_selected_Bcells, reduction = "umap", group.by = "immgen_SP.fine", split.by = "huIgk")

DimPlot(Seurat_Object_BM_selected_Bcells, reduction = "umap", group.by = "immgen_BM.fine", split.by = "huIgk")

DimPlot(Seurat_Object_BM_selected_Bcells, reduction = "umap", group.by = "seurat_clusters", split.by = "genotype", label = TRUE)

DimPlot(Seurat_Object_BM_selected_Bcells, reduction = "umap", group.by = "Phase", split.by = "genotype", label = TRUE)
```

```{r}
# Reorder levels manualy in the Seurat object
Seurat_Object_BM_selected_Bcells$immgen_BM.fine <- factor(Seurat_Object_BM_selected_Bcells$immgen_BM.fine, 
                                  levels = c("B cells, pro (proB.FrA)", 
                                             "B cells (proB.CLP)", 
                                             "B cells (proB.FrBC)", 
                                             "B cells (preB.FrC)", 
                                             "B cells (preB.FrD)", 
                                             "B cells (B.FrE)", 
                                             "B cells (B.FrF)"))
DimPlot(Seurat_Object_BM_selected_Bcells, group.by = "immgen_BM.fine")
```


# Histogram of B cell subtypes

```{r}
Seurat_Object_BM_A12 <- subset(Seurat_Object_BM_selected_Bcells, genotype == "A12")
Seurat_Object_BM_WT <- subset(Seurat_Object_BM_selected_Bcells, genotype == "WT")

### FOR WT ###

# Get all unique categories from Seurat_Object_BM_WT$immgen_BM.fine
cell_subtypes <- unique(Seurat_Object_BM_WT$immgen_BM.fine)

# Calculate the frequency of each category in Seurat_Object_BM_WT$immgen_BM.fine
freq_table_WT <- table(Seurat_Object_BM_WT$immgen_BM.fine)

# Convert the frequency table into a data frame and sort it from highest to lowest
freq_df_WT <- as.data.frame(freq_table_WT)
freq_df_WT <- freq_df_WT[order(-freq_df_WT$Freq), ]

# Create a bar plot with ggplot2 using colors based on the categories
bar_plot <- ggplot(freq_df_WT, aes(x = reorder(Var1, Freq), y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", color = "black", alpha = 0.7) +  # Bar plot
  scale_fill_viridis_d() +  # Continuous color scale (viridis)
  labs(title = "Frequency of B Cell Subtypes in BM (WT)", x = "B Cell Subtypes", y = "Frequency") +  # Title and axis labels
  theme_minimal() +  # Theme style
  theme(axis.text.x = element_text(angle = 90, hjust = 1))  # Rotate x-axis labels

# Display the bar plot
print(bar_plot)

### FOR A12 ###

# Get all unique categories from Seurat_Object_BM_A12$immgen_BM.fine
cell_subtypes <- unique(Seurat_Object_BM_A12$immgen_BM.fine)

# Calculate the frequency of each category in Seurat_Object_BM_A12$immgen_BM.fine
freq_table_A12 <- table(Seurat_Object_BM_A12$immgen_BM.fine)

# Convert the frequency table into a data frame and sort it from highest to lowest
freq_df_A12 <- as.data.frame(freq_table_A12)
freq_df_A12 <- freq_df_A12[order(-freq_df_A12$Freq), ]

# Create a bar plot with ggplot2 using colors based on the categories
bar_plot <- ggplot(freq_df_A12, aes(x = reorder(Var1, Freq), y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", color = "black", alpha = 0.7) +  # Bar plot
  scale_fill_viridis_d() +  # Continuous color scale (viridis)
  labs(title = "Frequency of B Cell Subtypes in BM (A12)", x = "B Cell Subtypes", y = "Frequency") +  # Title and axis labels
  theme_minimal() +  # Theme style
  theme(axis.text.x = element_text(angle = 90, hjust = 1))  # Rotate x-axis labels

# Display the bar plot
print(bar_plot)

```

```{r}
Seurat_Object_SP_A12 <- subset(Seurat_Object_SP_selected_Bcells, genotype == "A12")
Seurat_Object_SP_WT <- subset(Seurat_Object_SP_selected_Bcells, genotype == "WT")

### FOR WT ###

# Get all unique categories from Seurat_Object_SP_WT$immgen_SP.fine
cell_subtypes <- unique(Seurat_Object_SP_WT$immgen_SP.fine)

# Calculate the frequency of each category in Seurat_Object_SP_WT$immgen_SP.fine
freq_table_WT <- table(Seurat_Object_SP_WT$immgen_SP.fine)

# Convert the frequency table into a data frame and sort it from highest to lowest
freq_df_WT <- as.data.frame(freq_table_WT)
freq_df_WT <- freq_df_WT[order(-freq_df_WT$Freq), ]

# Create a bar plot with ggplot2 using colors based on the categories
bar_plot <- ggplot(freq_df_WT, aes(x = reorder(Var1, Freq), y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", color = "black", alpha = 0.7) +  # Bar plot
  scale_fill_viridis_d() +  # Continuous color scale (viridis)
  labs(title = "Frequency of B Cell Subtypes in SP (WT)", x = "B Cell Subtypes", y = "Frequency") +  # Title and axis labels
  theme_minimal() +  # Theme style
  theme(axis.text.x = element_text(angle = 90, hjust = 1))  # Rotate x-axis labels

# Display the bar plot
print(bar_plot)

### FOR A12 ###

# Get all unique categories from Seurat_Object_SP_A12$immgen_SP.fine
cell_subtypes <- unique(Seurat_Object_SP_A12$immgen_SP.fine)

# Calculate the frequency of each category in Seurat_Object_SP_A12$immgen_SP.fine
freq_table_A12 <- table(Seurat_Object_SP_A12$immgen_SP.fine)

# Convert the frequency table into a data frame and sort it from highest to lowest
freq_df_A12 <- as.data.frame(freq_table_A12)
freq_df_A12 <- freq_df_A12[order(-freq_df_A12$Freq), ]

# Create a bar plot with ggplot2 using colors based on the categories
bar_plot <- ggplot(freq_df_A12, aes(x = reorder(Var1, Freq), y = Freq, fill = Var1)) +
  geom_bar(stat = "identity", color = "black", alpha = 0.7) +  # Bar plot
  scale_fill_viridis_d() +  # Continuous color scale (viridis)
  labs(title = "Frequency of B Cell Subtypes in SP (A12)", x = "B Cell Subtypes", y = "Frequency") +  # Title and axis labels
  theme_minimal() +  # Theme style
  theme(axis.text.x = element_text(angle = 90, hjust = 1))  # Rotate x-axis labels

# Display the bar plot
print(bar_plot)
```

# Cellular cycle

```{r}
CC_BM <- CellCycleScoring(Seurat_Object_BM_selected_Bcells, s.features = cc.genes$s.genes, g2m.features = cc.genes$g2m.genes)
DimPlot(CC_BM, group.by = "Phase", split.by = "genotype")
```

```{r}
CC_SP <- CellCycleScoring(Seurat_Object_SP_selected_Bcells, s.features = cc.genes$s.genes, g2m.features = cc.genes$g2m.genes)
DimPlot(CC_SP, group.by = "Phase", split.by = "genotype")
```

# Assign celltype based on reference 

# Pre-ProB cell identification

```{r}
FeaturePlot(Seurat_Object_BM_WT, features = c("Il7r","Irf8","Tcf4","Bst2","Tyrobp","Clec12a","Cd300a","Mycl"), max.cutoff = 4,
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = c("Flt3", "Tyrobp","Irf8"), max.cutoff = 4,
    cols = c("grey", "red"))
```

```{r}
# List of genes for Pre-ProB signature
PrePro_marker_gene_list <- list(c("Flt3", "Tyrobp", "Irf8", "Il7r"))

# Calculate module for specified genes
Seurat_Object_BM_selected_Bcells <- AddModuleScore(
  object = Seurat_Object_BM_selected_Bcells, 
  features = PrePro_marker_gene_list, 
  name = "PrePro_score"
)

# Create feature Plot
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = "PrePro_score1") + 
  scale_color_viridis(option = "D") +  
  ggtitle("PreProB Score", subtitle = "Genes: Flt3, Tyrobp, Irf8, Il7r") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16),  
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )
```


# ProB

```{r}
FeaturePlot(Seurat_Object_BM_WT, features = c("Vpreb1", "Igll1","Bok","Ebf1","Ifitm3","Ifitm2"), max.cutoff = 4,
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = c("Vpreb1","Igll1","Ifitm2", "Bok"), split.by = "genotype", max.cutoff = 4,
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = c("Vpreb1","Igll1"), split.by = "genotype", max.cutoff = 4,
    cols = c("grey", "red"))
```

```{r}
# List of genes for ProB subset
Pro_marker_gene_list <- list(c("Vpreb1","Igll1","Ifitm2", "Bok", "Ifitm3"))

# Calculate module for specified genes
Seurat_Object_BM_selected_Bcells <- AddModuleScore(
  object = Seurat_Object_BM_selected_Bcells, 
  features = Pro_marker_gene_list, 
  name = "Pro_score"
)

# Create feature Plot
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = "Pro_score1") + 
  scale_color_viridis(option = "D") +  
  ggtitle("ProB Score", subtitle = "Genes: Vpreb1, Igll1, Ifitm2, Ifitm3, Bok") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16), 
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )
```

# Large PreB identification

```{r}
FeaturePlot(Seurat_Object_BM_WT, features = c("Mki67", "Nrgn"), max.cutoff = 4,
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = c("Mki67", "Nrgn", "Ybx3"), split.by = "genotype", max.cutoff = 4,
    cols = c("grey", "red"))
```

```{r}
# List of genes for Large PreB signature
LargePreB_marker_gene_list <- list(c("Mki67", "Nrgn", "Ybx3"))

# Calculate module for specified genes
Seurat_Object_BM_selected_Bcells <- AddModuleScore(
  object = Seurat_Object_BM_selected_Bcells, 
  features = LargePreB_marker_gene_list, 
  name = "LargePreB_score"
)

# Create feature Plot
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = "LargePreB_score1") + 
  scale_color_viridis(option = "D") +  
  ggtitle("LargePreB Score", subtitle = "Genes: Mki67, Nrgn, Ybx3") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16), 
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )
```

# Small preB gene signature

```{r}
FeaturePlot(Seurat_Object_BM_WT, features = c("Cd74", "Rag1", "Rag2"), max.cutoff = 4,
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = c("Cd74", "Rag1", "Rag2", "Igkc", "Cxcr4"), split.by = "genotype", max.cutoff = 4,
    cols = c("grey", "red"))
```

```{r}
# List of genes for small PreB
SmallPreB_marker_gene_list <- list(c("Rag1", "Rag2", "Bach2"))

# Calculate module for specified genes
Seurat_Object_BM_selected_Bcells <- AddModuleScore(
  object = Seurat_Object_BM_selected_Bcells, 
  features = SmallPreB_marker_gene_list, 
  name = "SmallPreB_score"
)

# Create feature Plot
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = "SmallPreB_score1") + 
  scale_color_viridis(option = "D") +  
  ggtitle("SmallPreB Score", subtitle = "Genes: Rag1, Rag2, Bach2") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16),  
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )
```

# Identificación Immature

```{r}
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = c("Igkc", "Iglc1", "Iglc2", "Iglc3"), split.by = "genotype", max.cutoff = 4,
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = c("Igkc"), split.by = "genotype", max.cutoff = 4,
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = c("Iglc3"), split.by = "genotype", max.cutoff = 4,
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = c("Rag1", "Rag2"), split.by = "genotype", max.cutoff = 4,
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = c("H2-Aa", "Ms4a4c", "Ms4a1"), split.by = "genotype", max.cutoff = 4,
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = c("Ighd"), split.by = "genotype", max.cutoff = 4,
    cols = c("grey", "red"))
```

```{r}
# List of genes for Immature
Immature_marker_gene_list <- list(c("CD74", "Ms4a1"))

# Calculate module for specified genes
Seurat_Object_BM_selected_Bcells <- AddModuleScore(
  object = Seurat_Object_BM_selected_Bcells, 
  features = Immature_marker_gene_list, 
  name = "Immature_score"
)

# Create feature plot
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = "Immature_score1") + 
  scale_color_viridis(option = "D") +  
  ggtitle("Immature Score", subtitle = "Genes: CD74, Ms4a1") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16),  
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )
```

```{r}
# List of genes for Mature B
Mature_marker_gene_list <- list(c("CD74", "H2-Aa", "Ms4a4c"))

#Calculate module for specified genes
Seurat_Object_BM_selected_Bcells <- AddModuleScore(
  object = Seurat_Object_BM_selected_Bcells, 
  features = Mature_marker_gene_list, 
  name = "Mature_score"
)

# Create Feature Plot
FeaturePlot(Seurat_Object_BM_selected_Bcells, features = "Mature_score1") + 
  scale_color_viridis(option = "D") +  
  ggtitle("Mature B Score", subtitle = "Genes: CD74, Ms4a4c, H2-Aa") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16),  
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )
```


```{r}
DimPlot(Seurat_Object_BM_selected_Bcells, split.by = "genotype")
DimPlot(Seurat_Object_BM_selected_Bcells, group.by = "immgen_BM.fine")
DimPlot(Seurat_Object_BM_selected_Bcells, group.by = "Phase")
```



# Rename clusters and analysis for BM

```{r}
DimPlot(Seurat_Object_BM_selected_Bcells, reduction = "umap", group.by = "immgen_BM.fine") +
    theme(
        text = element_text(family = "Times New Roman")  
    )
DimPlot(Seurat_Object_BM_selected_Bcells, reduction = "umap", group.by = "seurat_clusters", label = TRUE) +
    theme(
        text = element_text(family = "Times New Roman")
    )
```

```{r}
Seurat_Object_BM_selected_Bcells_idents <- RenameIdents(Seurat_Object_BM_selected_Bcells, "0"="Mature B", "1"="Immature B", "2"="Small PreB", "3"="Small PreB", "4"= "A12 unique", "5"="Pre-Pro B", "6"="Large PreB", "7"="Mature B", "8"="Large PreB", "9"="Cycling Pre-ProB", "10"="Large PreB", "11"= "Cycling ProB", "12" = "ProB", "13"="Cycling Immature", "14" = "Large PreB", "15" = "Mature B", "16" = "Mature B", "17" = "Highly Mitochondrial")
```

```{r}
Seurat_Object_BM_selected_Bcells_idents$seurat_clusters_new <- Idents(Seurat_Object_BM_selected_Bcells_idents)
```

```{r}
# Now create the DimPlot with specified colors
DimPlot(
    Seurat_Object_BM_selected_Bcells_idents, 
    reduction = "umap",
    group.by = "seurat_clusters_new",
    cols=  c(
    "Mature B" = "#E57373",       
    "Cycling Immature" = "#F4A582",
    "A12 unique" = "#DDA0DD",     
    "Immature B" = "#A39BD3",     
    "Small PreB" = "#D39BA3",     
    "Large PreB" = "#A9BBA9",      
    "ProB" = "#A93226",             
    "Cycling ProB" = "#E9BBA9",     
    "Pre-Pro B" = "#56A7C9",        
    "Cycling Pre-ProB" = "#4C59B6"  
)
)

DimPlot(
    Seurat_Object_BM_selected_Bcells_idents, 
    reduction = "umap",
    group.by = "seurat_clusters_new",
    cols=  c(
    "Mature B" = "#E57373",         
    "Cycling Immature" = "#F4A582", 
    "A12 unique" = "#DDA0DD",       
    "Immature B" = "#A39BD3",     
    "Small PreB" = "#D39BA3",       
    "Large PreB" = "#A9BBA9",      
    "ProB" = "#A93226",            
    "Cycling ProB" = "#E9BBA9",     
    "Pre-Pro B" = "#56A7C9",        
    "Cycling Pre-ProB" = "#4C59B6"  
), split.by = "genotype"
) +
    theme(
        text = element_text(family = "Times New Roman") 
    )
```



```{r}

dittoBarPlot(
    object = Seurat_Object_BM_selected_Bcells_idents,
    scale = "percent",
    var = "seurat_clusters_new",
    color.panel =  c(
    "Mature B" = "#E57373",         
    "Cycling Immature" = "#F4A582",
    "A12 unique" = "#DDA0DD",      
    "Immature B" = "#A39BD3",       
    "Small PreB" = "#D39BA3",       
    "Large PreB" = "#A9BBA9",      
    "ProB" = "#A93226",           
    "Cycling ProB" = "#E9BBA9",     
    "Pre-Pro B" = "#56A7C9",        
    "Cycling Pre-ProB" = "#4C59B6" 
), group.by = "genotype", x.reorder = c(2,1), var.labels.reorder = c(9,3,10,4,7,1,11,6,2,8,5))+
    theme(
        text = element_text(family = "Times New Roman")  
    )

```

```{r}

# Extract data from Seurat object
data <- Seurat_Object_BM_selected_Bcells_idents@meta.data
# Calculate percentages for each combination of genotype and seurat_clusters
data_summary <- data %>%
  group_by(genotype, seurat_clusters) %>%
  summarize(count = n()) %>%
  mutate(percentage = count / sum(count) * 100)
data_summary
```


```{r}
# Just a quick example with the built in Seurat data:
dittoBarPlot(
    object = Seurat_Object_BM_selected_Bcells_idents,
    scale = "percent",
    var = "seurat_clusters_new",
    color.panel = c(
    "Mature B" = "#E57373",         
    "Cycling Immature" = "#F4A582", 
    "A12 unique" = "#DDA0DD",      
    "Immature B" = "#A39BD3",       
    "Small PreB" = "#D39BA3",       
    "Large PreB" = "#A9BBA9",     
    "ProB" = "#A93226",             
    "Cycling ProB" = "#E9BBA9",    
    "Pre-Pro B" = "#56A7C9",        
    "Cycling Pre-ProB" = "#4C59B6"  
),
    group.by = "mice", x.reorder = c(4,5,6,1,2,3), var.labels.reorder = c(9,3,10,4,7,1,11,6,2,8,5))
```

# Rename clusters and analysis for SP


```{r}
DimPlot(Seurat_Object_SP_selected_Bcells, reduction = "umap", group.by = "immgen_SP.fine") +
    theme(
        text = element_text(family = "Times New Roman")  
    )
DimPlot(Seurat_Object_SP_selected_Bcells, reduction = "umap", group.by = "seurat_clusters", label = TRUE) +
    theme(
        text = element_text(family = "Times New Roman") 
    )
```


```{r}
Seurat_Object_SP_selected_Bcells_idents <- RenameIdents(Seurat_Object_SP_selected_Bcells, "0"="Fo B", "1"="Fo B", "2" = "MZ B", "3"="Transitional B", "4"= "MZ B", "5"="Mix", "6"="Transitional B", "7" = "B1", "8" = "Fo B", "9" = "GC B", "10" = "Fo B")
```


```{r}
Seurat_Object_SP_selected_Bcells_idents$seurat_clusters_new <- Idents(Seurat_Object_SP_selected_Bcells_idents)
```


```{r}
#plot the cells from Blimp using the transferred clustering
DimPlot(Seurat_Object_SP_selected_Bcells_idents, reduction = "umap", cols= c("Fo B" = "#D55E00", "Transitional B" = "darkred", "MZ B" = "#009E73", "B1" = "#CC79A7", "Mix" = "darkslategray", "GC B" = "tan4"), group.by = "seurat_clusters_new")
DimPlot(Seurat_Object_SP_selected_Bcells_idents, reduction = "umap", cols= c("Fo B" = "#D55E00", "Transitional B" = "darkred", "MZ B" = "#009E73", "B1" = "#CC79A7", "Mix" = "darkslategray", "GC B" = "tan4"), group.by = "seurat_clusters_new", split.by = "genotype") +
    theme(
        text = element_text(family = "Times New Roman")  # Cambia la fuente a Times New Roman
    )
```

```{r}
data_summary <- Seurat_Object_SP_selected_Bcells_idents@meta.data %>%
  group_by(genotype, seurat_clusters_new) %>%
  summarize(count = n()) %>%
  mutate(percentage = count / sum(count) * 100)

data_summary[data_summary$genotype == "A12",]
data_summary[data_summary$genotype == "WT",]
```

```{r}
DimPlot(Seurat_Object_SP_selected_Bcells_idents, reduction = "umap", group.by = "immgen_SP.fine") + DimPlot(Seurat_Object_SP_selected_Bcells_idents, reduction = "umap", group.by = "seurat_clusters", label = T)
```


```{r}
Seurat_Object_SP_selected_Bcells_idents$seurat_clusters_new<- Idents(Seurat_Object_SP_selected_Bcells_idents)
```


```{r}
# Just a quick example with the built in Seurat data:
dittoBarPlot(
    object = Seurat_Object_SP_selected_Bcells_idents,
    scale = "percent",
    var = "seurat_clusters_new",
    color.panel =  c("Fo B" = "#D55E00", "Transitional B" = "darkred", "MZ B" = "#009E73", "B1" = "#CC79A7", "Mix" = "darkslategray", "GC B" = "tan4"),
    group.by = "genotype", x.reorder = c(2,1))
```

```{r}
# Just a quick example with the built in Seurat data:
dittoBarPlot(
    object = Seurat_Object_SP_selected_Bcells_idents,
    scale = "percent",
    var = "seurat_clusters_new",
    color.panel = c("Fo B" = "#D55E00", "Transitional B" = "darkred", "MZ B" = "#009E73", "B1" = "#CC79A7", "Mix" = "darkslategray", "GC B" = "tan4"),
    group.by = "mice", x.reorder = c(4,5,6,1,2,3))
```

### Gene Expression

# A12 expression

```{r}
FeaturePlot(Seurat_Object_BM_selected_Bcells_idents, features = "A12", max.cutoff = 1, split.by = "genotype",
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells_idents, features = "A12", max.cutoff = 1, split.by = "mice",
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells_idents, features = "A12", max.cutoff = 1, split.by = "huIgk",
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells_idents, features = "A12", max.cutoff = 3,
    cols = c("grey", "red"))
DimPlot(Seurat_Object_BM_selected_Bcells_idents, group.by = "immgen_BM.fine", split.by = "genotype")
```

```{r}
FeaturePlot(Seurat_Object_SP_selected_Bcells_idents, features = "A12", max.cutoff = 1, split.by = "genotype",
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells_idents, features = "Rag1", max.cutoff = 1, split.by = "genotype",
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells_idents, features = "Rag2", max.cutoff = 1, split.by = "genotype",
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells_idents, features = "Igkc", max.cutoff = 1, split.by = "genotype",
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_SP_selected_Bcells_idents, features = "A12", max.cutoff = 1, split.by = "mice",
    cols = c("grey", "red"))
```
# Expresión de CD21 y CD23
```{r}
FeaturePlot(Seurat_Object_SP_selected_Bcells_idents, features = "Fcer2a", max.cutoff = 1, split.by = "genotype",
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_SP_selected_Bcells_idents, features = "Il10ra", max.cutoff = 1, split.by = "genotype",
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_BM_selected_Bcells_idents, features = "Il7r", max.cutoff = 5, split.by = "genotype",
    cols = c("grey", "red"))
FeaturePlot(Seurat_Object_SP_selected_Bcells_idents, features = "Aldh3b1", max.cutoff = 2, split.by = "genotype",
    cols = c("grey", "red"))
```

# Most Relevant Genes per Cluster

# Generate subsets for A12 and WT

```{r}
Seurat_Object_BM_A12 <- subset(Seurat_Object_BM_selected_Bcells_idents, genotype == "A12")
Seurat_Object_BM_WT <- subset(Seurat_Object_BM_selected_Bcells_idents, genotype == "WT")
Seurat_Object_SP_A12 <- subset(Seurat_Object_SP_selected_Bcells_idents, genotype == "A12")
Seurat_Object_SP_WT <- subset(Seurat_Object_SP_selected_Bcells_idents, genotype == "WT")
```

The min.pct argument requires a feature to be detected at a minimum percentage in either of the two groups of cells. We report only the positive ones.  

# Wilcox spleen WT analysis
```{r}
all_markers_SP <- FindAllMarkers(object = Seurat_Object_SP_WT, only.pos = T, min.pct = 0.25, min.diff.pct = 0.1, thresh.use = 0.25)
```
We print the two most relevant per cluster:
```{r}
five_top_markers_SP <- all_markers_SP %>%
                        group_by(cluster) %>%
                        dplyr::filter(avg_log2FC > 1 & (pct.1 > 0.25 | pct.2 > 0.25) & p_val_adj < 1e-20) %>%
                        slice_min(n = 5, order_by = p_val_adj) %>%
                        ungroup() -> top5_SP

five_top_markers_SP
DoHeatmap(Seurat_Object_SP_WT, features = top5_SP$gene)

average_expression <- AverageExpression(Seurat_Object_SP_WT, features = five_top_markers_SP$gene, return.seurat = TRUE)

DoHeatmap(average_expression, features = five_top_markers_SP$gene, size = 3, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red"))
```

# Wilcox WT Bone Marrow Analysis
```{r}
all_markers_BM <- FindAllMarkers(object = Seurat_Object_BM_WT, only.pos = T, min.pct = 0.25, min.diff.pct = 0.1, thresh.use = 0.25)
```
We print the five most relevant per cluster:
```{r}
five_top_markers_BM <- all_markers_BM %>%
                        group_by(cluster) %>%
                        dplyr::filter(avg_log2FC > 1 & (pct.1 > 0.3 | pct.2 > 0.3) & p_val_adj < 1e-80) %>%
                        slice_max(n = 5, order_by = avg_log2FC) %>%
                        ungroup() -> top5_BM

five_top_markers_BM
DoHeatmap(Seurat_Object_BM_WT, features = top5_BM$gene)

average_expression <- AverageExpression(Seurat_Object_BM_WT, features = five_top_markers_BM$gene, return.seurat = TRUE)

DoHeatmap(average_expression, features = five_top_markers_BM$gene, size = 3, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red"))
```

```{r}
twenty_top_markers_BM <- all_markers_BM %>%
                        group_by(cluster) %>%
                        dplyr::filter(avg_log2FC > 1 & (pct.1 > 0.3 | pct.2 > 0.3) & p_val_adj < 1e-30) %>%
                        slice_max(n = 20, order_by = avg_log2FC) %>%
                        ungroup() -> top20_BM

twenty_top_markers_BM
DoHeatmap(Seurat_Object_BM_WT, features = top20_BM$gene)

# Calcular la expresión promedio por cluster para los genes de interés
average_expression <- AverageExpression(Seurat_Object_BM_WT, features = top20_BM$gene, return.seurat = TRUE)

# Crear el heatmap con los datos de expresión promedio
DoHeatmap(average_expression, features = top20_BM$gene, size = 3, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red"))
```

### Subclustering B1

# Redo Scaling, PCA and clustering for B1

## Identifucation of highly variable features of B1

```{r}
Seurat_Object_B1_cells <- subset(Seurat_Object_SP_selected_Bcells_idents, seurat_clusters_new == "B1")

Seurat_Object_B1_cells <- FindVariableFeatures(Seurat_Object_B1_cells, selection.method = "vst", nfeatures = 3000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(Seurat_Object_B1_cells), 10)

# plot variable features with labels
plot1 <- VariableFeaturePlot(Seurat_Object_B1_cells)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```

```{r}
variable_genes <- setdiff(VariableFeatures(Seurat_Object_B1_cells), c("A12"))
```


## Scale B1 cells

```{r}
Seurat_Object_B1_cells <- ScaleData(Seurat_Object_B1_cells, features = VariableFeatures(object = Seurat_Object_B1_cells))
```

## PCA of B1 cells

```{r}
Seurat_Object_B1_cells <- RunPCA(Seurat_Object_B1_cells, features = variable_genes)
```

Representation:
```{r}
DimPlot(Seurat_Object_B1_cells, reduction = "pca")
```


### Determine the ‘dimensionality’ of the dataset

```{r}
#Seurat_Object_B1_cells <- JackStraw(Seurat_Object_B1_cells, num.replicate = 100, dims = 40)
```

```{r}
#Seurat_Object_B1_cells <- ScoreJackStraw(Seurat_Object_B1_cells, dims = 1:40)
```

```{r}
#JackStrawPlot(Seurat_Object_B1_cells, dims = 1:40)
```

```{r}
ElbowPlot(Seurat_Object_B1_cells, ndims=40)
```


## B1 cells clusters

```{r}
Seurat_Object_B1_cells <- FindNeighbors(Seurat_Object_B1_cells, dims = 1:10)
Seurat_Object_B1_cells <- FindClusters(Seurat_Object_B1_cells, resolution = 0.3)
```


```{r}
Seurat_Object_B1_cells <- RunUMAP(Seurat_Object_B1_cells, dims = 1:10)
DimPlot(Seurat_Object_B1_cells, reduction = "umap", group.by = "mice")
DimPlot(Seurat_Object_B1_cells, reduction = "umap", group.by = "genotype")
DimPlot(Seurat_Object_B1_cells, reduction = "umap", group.by = "seurat_clusters")
FeaturePlot(Seurat_Object_B1_cells, features = "A12")
FeaturePlot(Seurat_Object_B1_cells, features = "Ssbp2")
DimPlot(Seurat_Object_SP_selected_Bcells_idents, split.by = "genotype")
```


# Rename clusters and analysis for SP

```{r}
DimPlot(Seurat_Object_B1_cells, reduction = "umap", group.by = "genotype") + DimPlot(Seurat_Object_B1_cells, reduction = "umap", group.by = "seurat_clusters", label = TRUE, cols = c("darkblue", "violet"))
```

```{r}
Seurat_Object_B1_cells_idents <- RenameIdents(Seurat_Object_B1_cells, "1" = "A12 expressing", "0" = "B1")
Seurat_Object_B1_cells_idents$seurat_clusters_new_fin <- Idents(Seurat_Object_B1_cells_idents)

```

```{r}
#plot the cells from Blimp using the transferred clustering
DimPlot(Seurat_Object_B1_cells_idents, reduction = "umap", cols= c("A12 expressing" = "#CC14A7", "B1" = "darkblue"), group.by = "seurat_clusters_new_fin")
DimPlot(Seurat_Object_B1_cells_idents, reduction = "umap", cols= c("A12 expressing" = "#CC14A7",  "B1" = "#CC79A7"), split.by = "genotype", group.by = "seurat_clusters_new_fin")
```

```{r}
all_markers_B1_subcl <- FindAllMarkers(object = Seurat_Object_B1_cells_idents, only.pos = T, min.pct = 0.25, min.diff.pct = 0.1, thresh.use = 0.25)
```
We print the five most relevant per cluster:
```{r}
five_top_markers_B1_subcl <- all_markers_B1_subcl %>%
                        group_by(cluster) %>%
                        dplyr::filter(avg_log2FC > 1 & (pct.1 > 0.3 | pct.2 > 0.3) & p_val_adj < 1e-4) %>%
                        slice_max(n = 20, order_by = avg_log2FC) %>%
                        ungroup() 

five_top_markers_B1_subcl
DoHeatmap(Seurat_Object_B1_cells_idents, features = five_top_markers_B1_subcl$gene)

average_expression <- AverageExpression(Seurat_Object_B1_cells_idents, features = five_top_markers_B1_subcl$gene, return.seurat = TRUE)

DoHeatmap(average_expression, features = five_top_markers_B1_subcl$gene, size = 3, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red"))
```

# Rename in original clusters

```{r}
A12_expressing_cells <- Cells(Seurat_Object_B1_cells_idents)[which(Seurat_Object_B1_cells_idents$seurat_clusters == "1")]
```

```{r}
Idents(Seurat_Object_SP_selected_Bcells_idents, cells = A12_expressing_cells) <- "A12 expressing"
```

```{r}
Seurat_Object_SP_selected_Bcells_idents$seurat_clusters_new_fin <- Idents(Seurat_Object_SP_selected_Bcells_idents)
```

```{r}
DimPlot(Seurat_Object_SP_selected_Bcells_idents, reduction = "umap", split.by = "genotype", group.by = "seurat_clusters_new_fin")
```

```{r}
# Just a quick example with the built in Seurat data:
DimPlot(Seurat_Object_SP_selected_Bcells_idents, reduction = "umap", cols= c("Fo B" = "#D55E00", "Transitional B" = "darkred", "MZ B" = "#009E73", "A12 expressing" = "#DC79A7", "Mix" = "darkslategray", "GC B" = "tan4", "B1" = "#ED8987A7"), group.by = "seurat_clusters_new_fin", split.by = "genotype") +
    theme(
        text = element_text(family = "Times New Roman") 
    )

# Just a quick example with the built in Seurat data:
DimPlot(Seurat_Object_SP_selected_Bcells_idents, reduction = "umap", cols= c("Fo B" = "#D55E00", "Transitional B" = "darkred", "MZ B" = "#009E73", "A12 expressing" = "#DC79A7", "Mix" = "darkslategray", "GC B" = "tan4", "B1" = "#CD9987A7"), group.by = "seurat_clusters_new_fin", split.by = "genotype") +
    theme(
        text = element_text(family = "Times New Roman") 
    )


dittoBarPlot(
    object = Seurat_Object_SP_selected_Bcells_idents,
    scale = "percent",
    var = "seurat_clusters_new_fin",
    color.panel = c("Fo B" = "#D55E00", "Transitional B" = "darkred", "MZ B" = "#009E73", "A12 expressing" = "#CC79A7", "Mix" = "darkslategray", "GC B" = "tan4", "B1" = "#CD9987A7"),
    group.by = "genotype", x.reorder = c(2,1))+
    theme(
        text = element_text(family = "Times New Roman")  
    )
```

# Statistics t-test for SP

```{r}
# Obtain percentages for each cluster in each mouse
data_summary <- Seurat_Object_SP_selected_Bcells_idents@meta.data %>%
  group_by(genotype, mice, seurat_clusters_new_fin) %>%
  summarize(count = n()) %>%
  mutate(percentage = count / sum(count) * 100)

# Store percentages in lists for each genotype
A12_test <- list()
WT_test <- list()
for (i in unique(data_summary$seurat_clusters_new_fin)) {
  A12_test[[as.character(i)]] <- data_summary[data_summary$genotype == "A12" & data_summary$seurat_clusters_new_fin == i, ]
  WT_test[[as.character(i)]] <- data_summary[data_summary$genotype == "WT" & data_summary$seurat_clusters_new_fin == i, ]
}

# Create an empty list to store t-test results
t_test_results <- list()

# Perform t-tests for each cluster between genotypes
for (i in names(A12_test)) {
  group1 <- A12_test[[i]]$percentage
  group2 <- WT_test[[i]]$percentage
  
  # Perform the t-test
  result <- t.test(group1, group2)
  
  # Store the results in the list
  t_test_results[[paste("Comparison", i)]] <- result
}

# Rename the result names to simplify
names(t_test_results) <- sub("Comparison ", "", names(t_test_results))

# Extract p-values from the t-test results
p_values <- sapply(t_test_results, function(res) res$p.value)
p_values

# Calculate log fold change and add significance markers
mean_per_cluster_A12 <- sapply(A12_test, function(df) mean(df$percentage, na.rm = TRUE))
mean_per_cluster_WT <- sapply(WT_test, function(df) mean(df$percentage, na.rm = TRUE))

# Extract p-values
p_value_list <- unlist(lapply(t_test_results, function(res) res$p.value))

# Calculate Log Fold Change (LFC)
log_fold_change <- log2(mean_per_cluster_A12 / mean_per_cluster_WT)

# Create a data frame with the results
results_df <- data.frame(
  cluster = names(A12_test),
  mean_A12 = mean_per_cluster_A12,
  mean_WT = mean_per_cluster_WT,
  log_fold_change = log_fold_change,
  p_value = p_value_list
)

# Determine levels of significance or include the p-value if between 0.05 and 0.1
results_df$significance <- ifelse(
  results_df$p_value <= 0.001, "***",
  ifelse(
    results_df$p_value <= 0.01, "**",
    ifelse(
      results_df$p_value <= 0.05, "*",
      ifelse(
        results_df$p_value <= 0.1,
        sprintf("%.3f", results_df$p_value), # Format to display p-value with 3 decimals
        ""
      )
    )
  )
)

print(results_df)

# Add significance levels to the summary for use in the plot
data_summary <- data_summary %>%
  left_join(results_df %>% select(cluster, significance), by = c("seurat_clusters_new_fin" = "cluster"))

# Create the plot with log fold change and significance (asterisks or p-values)
ggplot(data_summary, aes(x = seurat_clusters_new_fin, y = percentage, color = genotype)) +
  geom_point(size = 3, position = position_dodge(width = 0.5), alpha = 0.7) + 
  geom_text(
    data = results_df,
    aes(x = cluster, y = max(data_summary$percentage) + 5, label = significance), # Significance or p-value
    inherit.aes = FALSE, 
    size = 4,
    family = "Times New Roman" 
  ) +
  scale_color_manual(values = c("WT" = "black", "A12" = "violet")) + # Custom colors
  labs(
    x = "Cluster",
    y = "Percentage",
    color = "Genotype"
  ) +
  theme_minimal() +
  theme(
    text = element_text(family = "Times New Roman"),
    plot.title = element_text(size = 16, hjust = 0.5), 
    axis.text.x = element_text(size = 11, angle = 45, hjust = 1, family = "Times New Roman"),
    axis.text.y = element_text(size = 11, family = "Times New Roman"),
    strip.text = element_text(size = 18, family = "Times New Roman"),
    legend.text = element_text(size = 11, family = "Times New Roman"), 
    legend.title = element_text(family = "Times New Roman"), 
    legend.position = "bottom" 
  )

```


# Check difference between A12 expressing and non expressing

```{r}
Seurat_Object_FrBC <- subset(Seurat_Object_BM_selected_Bcells_idents, seurat_clusters_new %in% c("ProB", "Large PreB", "Cycling ProB"))
# Create new metadata category to indicate if they express A12 or not
Seurat_Object_FrBC$A12_status <- ifelse(Seurat_Object_FrBC@assays$RNA$counts["A12", ] > 1, "A12_expressing", "A12_non_expressing")

table(Seurat_Object_FrBC$A12_status)

# Find relevant markers
all_markers_A12_status <- FindMarkers(
  object = Seurat_Object_FrBC,
  ident.1 = "A12_expressing",
  ident.2 = "A12_non_expressing",
  group.by = "A12_status",
  only.pos = TRUE,
  min.pct = 0.25,
  min.diff.pct = 0.1,
  thresh.use = 0.25
)

# Filter 15 most relevant genes
five_top_markers_A12_status <- all_markers_A12_status %>%
  rownames_to_column("gene") %>%
  dplyr::filter(avg_log2FC > 1 & (pct.1 > 0.3 | pct.2 > 0.3) & p_val_adj < 1e-6) %>%
  slice_min(n = 15, order_by = p_val_adj)

print(five_top_markers_A12_status)

# Heatmap
DoHeatmap(Seurat_Object_FrBC, features = five_top_markers_A12_status$gene, group.by = "A12_status") +
  scale_fill_gradientn(colors = c("grey", "red"))

# Average expression
average_expression_A12 <- AverageExpression(
  Seurat_Object_FrBC,
  features = five_top_markers_A12_status$gene,
  group.by = "A12_status",
  return.seurat = TRUE
)

DoHeatmap(average_expression_A12, features = five_top_markers_A12_status$gene, size = 3, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red")) +
  labs(title = "Heatmap of Top Markers: A12 Expressing vs Non-Expressing")

```


```{r}
DimPlot(Seurat_Object_BM_selected_Bcells_idents, reduction = "umap") + DimPlot(Seurat_Object_BM_selected_Bcells_idents, reduction = "umap", group.by = "seurat_clusters", label = T) 
DimPlot(Seurat_Object_BM_selected_Bcells_idents, reduction = "umap", split.by = "genotype", group.by = "seurat_clusters_new")
DimPlot(Seurat_Object_BM_selected_Bcells_idents, reduction = "umap", split.by = "genotype", cells.highlight = Cells(Seurat_Object_BM_selected_Bcells_idents)[which(Seurat_Object_BM_selected_Bcells_idents$percent.mt > 4)])
```

# Bone Marrow Cluster Simplification

```{r}
DimPlot(Seurat_Object_BM_selected_Bcells_idents)

Seurat_Object_BM_selected_Bcells_idents$simplified_clusters <- "NA"
 
FrA_cells <- Cells(Seurat_Object_BM_selected_Bcells_idents)[which(Seurat_Object_BM_selected_Bcells_idents$seurat_clusters_new %in% c("Cycling Pre-ProB","Pre-Pro B"))]
Seurat_Object_BM_selected_Bcells_idents@meta.data[FrA_cells, "simplified_clusters"] = "Pre-Pro B"

FrB_cells <- Cells(Seurat_Object_BM_selected_Bcells_idents)[which(Seurat_Object_BM_selected_Bcells_idents$seurat_clusters_new %in% c("ProB", "Cycling ProB"))]
Seurat_Object_BM_selected_Bcells_idents@meta.data[FrB_cells, "simplified_clusters"] = "ProB"

FrD_cells <- Cells(Seurat_Object_BM_selected_Bcells_idents)[which(Seurat_Object_BM_selected_Bcells_idents$seurat_clusters_new %in% c("Small PreB"))]
Seurat_Object_BM_selected_Bcells_idents@meta.data[FrD_cells, "simplified_clusters"] = "Small PreB"

A12_unique <- Cells(Seurat_Object_BM_selected_Bcells_idents)[which(Seurat_Object_BM_selected_Bcells_idents$seurat_clusters_new %in% c("A12 unique"))]
Seurat_Object_BM_selected_Bcells_idents@meta.data[A12_unique, "simplified_clusters"] = "A12 unique"

FrF_cells <- Cells(Seurat_Object_BM_selected_Bcells_idents)[which(Seurat_Object_BM_selected_Bcells_idents$seurat_clusters_new %in% c("Mature B"))]
Seurat_Object_BM_selected_Bcells_idents@meta.data[FrF_cells, "simplified_clusters"] = "Mature B"

FrE_cells <- Cells(Seurat_Object_BM_selected_Bcells_idents)[which(Seurat_Object_BM_selected_Bcells_idents$seurat_clusters_new %in% c("Immature B", "Cycling Immature"))]
Seurat_Object_BM_selected_Bcells_idents@meta.data[FrE_cells, "simplified_clusters"] = "Immature B"

FrC_cells <- Cells(Seurat_Object_BM_selected_Bcells_idents)[which(Seurat_Object_BM_selected_Bcells_idents$seurat_clusters_new %in% c("Large PreB"))]
Seurat_Object_BM_selected_Bcells_idents@meta.data[FrC_cells, "simplified_clusters"] = "Large PreB"


```

Data visualization (umaps with clusters and barplots)

```{r}
Seurat_Object_BM_selected_Bcells_idents$simplified_clusters <-
  factor(Seurat_Object_BM_selected_Bcells_idents$simplified_clusters, 
         levels = c("Pre-Pro B", "ProB", "Large PreB", "A12 unique",
                    "Small PreB", "Immature B", "Mature B"))
Seurat_Object_BM_selected_Bcells_idents$genotype <- factor(Seurat_Object_BM_selected_Bcells_idents$genotype, levels = c("WT", "A12"))

DimPlot(Seurat_Object_BM_selected_Bcells_idents, group.by = "simplified_clusters", split.by = "genotype", cols= c("Immature B" = "tan" , "Mature B" = "tan2", "A12 unique" = "#E694C1", "Small PreB" = "lightblue3", "ProB" = "#A52A2A", "Pre-Pro B" = "palevioletred4", "Large PreB" = "#A694C1"))
DimPlot(Seurat_Object_BM_selected_Bcells_idents, group.by = "simplified_clusters", cols= c("Immature B" = "tan" , "Mature B" = "tan2", "A12 unique" = "#E694C1", "Small PreB" = "lightblue3", "ProB" = "#A52A2A", "Pre-Pro B" = "palevioletred4", "Large PreB" = "#A694C1")) + theme(text = element_text(family = "Times New Roman"))

Seurat_Object_BM_selected_Bcells_idents$simplified_clusters <-
  factor(Seurat_Object_BM_selected_Bcells_idents$simplified_clusters, 
         levels = c("Pre-Pro B", "ProB", "Large PreB", "A12 unique",
                    "Small PreB", "Immature B", "Mature B"))
dittoBarPlot(
  object = Seurat_Object_BM_selected_Bcells_idents,
  scale = "percent",
  var = "simplified_clusters",
  color.panel = c("Immature B" = "tan" , "Mature B" = "tan2", "A12 unique" = "#E694C1", "Small PreB" = "lightblue3", "ProB" = "#A52A2A", "Pre-Pro B" = "palevioletred4", "Large PreB" = "#A694C1"),
  group.by = "genotype", var.labels.reorder = c(5,6,3,1,7,2,4), x.reorder = c(2,1)
)
```


# Gene expression but taking into account new clusters

# Wilcox test of Spleen
```{r}
Seurat_Object_BM_A12 <- subset(Seurat_Object_BM_selected_Bcells_idents, genotype == "A12")
Seurat_Object_BM_WT <- subset(Seurat_Object_BM_selected_Bcells_idents, genotype == "WT")
Seurat_Object_SP_A12 <- subset(Seurat_Object_SP_selected_Bcells_idents, genotype == "A12")
Seurat_Object_SP_WT <- subset(Seurat_Object_SP_selected_Bcells_idents, genotype == "WT")

Idents(Seurat_Object_BM_A12) <- Seurat_Object_BM_A12$simplified_clusters
Idents(Seurat_Object_BM_WT) <- Seurat_Object_BM_WT$simplified_clusters

all_markers_SP <- FindAllMarkers(object = Seurat_Object_SP_A12, only.pos = T, min.pct = 0.25, min.diff.pct = 0.1, thresh.use = 0.25)
```
We print the two most relevant per cluster:
```{r}
five_top_markers_SP <- all_markers_SP %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1 & (pct.1 > 0.25 | pct.2 > 0.25) & p_val_adj < 1e-20) %>%
  slice_min(n = 20, order_by = p_val_adj) %>%
  ungroup() -> top5_SP

DoHeatmap(Seurat_Object_SP_A12, features = five_top_markers_SP$gene, size = 4, angle = 50)

average_expression <- AverageExpression(Seurat_Object_SP_A12, features = five_top_markers_SP$gene, return.seurat = TRUE)

DoHeatmap(average_expression, features = five_top_markers_SP$gene, group.colors = c("Fo B" = "#D55E00", "Transitional B" = "darkred", "MZ B" = "#009E73", "A12 expressing" = "#DC79A7", "Mix" = "darkslategray", "GC B" = "tan4", "B1" = "#ED8987A7"), size = 4, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red"))+
  theme(
    axis.text.y = element_text(size = 6), 
    plot.title = element_text(size = 12, family = "Times New Roman"),
    plot.subtitle = element_text(size = 15, family = "Times New Roman"),
    axis.title = element_text(size = 12, family = "Times New Roman"), 
    axis.text = element_text(size = 11, family = "Times New Roman") 
  )
```


```{r}
# Filter significant markers
five_top_markers_SP <- all_markers_SP %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1 & (pct.1 > 0.25 | pct.2 > 0.25) & p_val_adj < 1e-20) %>%
  slice_min(n = 5, order_by = p_val_adj) %>%
  ungroup() -> top5_SP

# Check top markers
print(five_top_markers_SP)

# Calculate average expression for selected genes
average_expression <- AverageExpression(Seurat_Object_SP_A12, features = five_top_markers_SP$gene)$RNA %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(cols = -gene, names_to = "cluster", values_to = "expression")

# Combine expression and p-values
data_for_dotplot <- five_top_markers_SP %>%
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj) %>%
  left_join(average_expression, by = c("gene", "cluster")) %>%
  mutate(
    p_val_adj = ifelse(p_val_adj == 0, 1e-300, p_val_adj), 
    significance = -log10(p_val_adj), 
    cluster = factor(cluster, levels = unique(cluster)) 
  )
# Create dotplot
# Normalize data of expression
data_for_dotplot <- data_for_dotplot %>%
  mutate(expression_scaled = log1p(expression)) # log1p(x) = log(1 + x)

ggplot(data_for_dotplot, aes(x = cluster, y = gene)) +
  geom_point(aes(size = significance, color = expression_scaled)) +
  scale_color_gradientn(
    colors = c("grey", "red"), 
    name = "Expression",
    limits = c(0, 3) 
  ) +
  scale_size_continuous(range = c(2, 8), name = "-log10(p-value)") +
  labs(
    x = "Cluster",
    y = "Gene",
    title = "Dot plot of gene expression and significance",
    subtitle = "Size indicates -log10(p-value); Color indicates log-scaled expression"
  ) +
  expand_limits(y = c(-0.5, length(unique(data_for_dotplot$gene)) + 0.5)) +  
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1),
    legend.position = "right"
  )

```



# Calculate percentajes of CITE-seq huIgk+ cells and highlight cells in the umap

```{r}
cells_to_highlight <- WhichCells(Seurat_Object_SP_selected_Bcells_idents, 
                                 expression = huIgk %in% c("huk_WT", "huk_A12"))

DimPlot(Seurat_Object_SP_selected_Bcells_idents, 
        group.by = "seurat_clusters", 
        cells.highlight = cells_to_highlight, split.by = "genotype")

positive_cells <- WhichCells(Seurat_Object_SP_A12, 
                             expression = huIgk %in% c("huk_WT", "huk_A12"))

Seurat_Object_SP_A12$huIgk_positive <- ifelse(
  Cells(Seurat_Object_SP_A12) %in% positive_cells, 
  "Positive", 
  "Negative"
)


meta_data <- Seurat_Object_SP_A12@meta.data

percentages <- meta_data %>%
  group_by(seurat_clusters_new_fin, huIgk_positive) %>%
  summarise(n = n()) %>%
  mutate(percentage = n / sum(n) * 100)

print(percentages)

ggplot(percentages, aes(x = seurat_clusters_new_fin, y = percentage, fill = huIgk_positive)) +
  geom_bar(stat = "identity", position = "stack") +  
  labs(title = "Frequence of positive cells per cluster",
       x = "Cluster",
       y = "Percentage",
       fill = "huIgk state") +
  theme_minimal()
```

# ROC analysis in SP for A12 expressing population
```{r}
all_markers_A12_uniq_pop <- FindMarkers(object = Seurat_Object_SP_A12, test.use = "roc", ident.1 = "A12 expressing", min.pct = 0.25, min.diff.pct = 0.1)
```
We print the two most relevant per cluster:
```{r}
five_top_markers_A12_uniq_pop_SP <- all_markers_A12_uniq_pop %>%
                        dplyr::filter((avg_log2FC > 1 | avg_log2FC < 1) & (power > 0.60)  & 
                                      (pct.1 > 0.25 | pct.2 > 0.25)) %>%
                        slice_max(n = 30, order_by = power) %>%
                        ungroup()

five_top_markers_A12_uniq_pop_SP 
DoHeatmap(Seurat_Object_SP_A12, features =rownames(five_top_markers_A12_uniq_pop_SP ), size = 4, angle = 50)

average_expression <- AverageExpression(Seurat_Object_SP_A12, features = five_top_markers_A12_uniq_pop_SP$gene, return.seurat = TRUE)

DoHeatmap(average_expression, features = row.names(five_top_markers_A12_uniq_pop_SP), size = 2, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red"))
```

Dotplot

```{r}
ggplot(five_top_markers_A12_uniq_pop_SP, aes(x = avg_log2FC, y = power, label = row.names(five_top_markers_A12_uniq_pop_SP))) +
  geom_point(aes(color = myAUC), alpha = 0.7, size = 4) +
  geom_text_repel() + # Etiquetas de genes
  scale_y_continuous(limits = c(0, 1), name = "Power") + 
  scale_color_viridis_c(option = "D", name = "AUC", limits = c(0, 1)) + 
  labs(
    x = "Average log2FC"
  ) +
  theme_minimal() +
  theme(
    text = element_text(family = "Times New Roman"),
    plot.title = element_text(size = 16, face = "bold", hjust = 0.5),
    axis.text = element_text(size = 12),
    axis.title = element_text(size = 14)
  )
```

# Mast analysis for A12 expressing in spleen

```{r}
all_markers_A12_uniq_pop <- FindMarkers(object = Seurat_Object_SP_A12, test.use = "MAST", ident.1 = "A12 expressing", min.pct = 0.25, min.diff.pct = 0.1)
```

DotPlot and Heatmap
```{r}
five_top_markers_A12_uniq_pop_SP <- all_markers_A12_uniq_pop %>%
  dplyr::filter((avg_log2FC > 1 | avg_log2FC < 1) & p_val_adj < 1e-30 & 
                                      (pct.1 > 0.25 | pct.2 > 0.25)) %>%
  slice_min(n = 30, order_by = p_val_adj) %>%
  ungroup() %>%
  rownames_to_column("gene")

average_expression <- AverageExpression(Seurat_Object_SP_A12, features = five_top_markers_A12_uniq_pop_SP$gene)$RNA %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(cols = -gene, names_to = "cluster", values_to = "expression")

DoHeatmap(Seurat_Object_SP_A12, features = average_expression$gene, size = 2, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red"))

data_for_dotplot <- five_top_markers_A12_uniq_pop_SP %>%
  dplyr::select(gene, avg_log2FC, p_val_adj) %>%
  left_join(average_expression, by = "gene") %>%
  mutate(
    p_val_adj = ifelse(p_val_adj == 0, 1e-300, p_val_adj), 
    significance = -log10(p_val_adj) 
  )

data_for_dotplot_A12 <- data_for_dotplot %>%
  arrange(significance) %>%
  mutate(gene = factor(gene, levels = unique(gene)))  

ggplot(data_for_dotplot_A12, aes(x = gene, y = cluster)) + 
  geom_point(aes(size = significance, color = avg_log2FC)) + 
  scale_color_gradientn(
    colors = c("blue", "white", "red"),  
    values = scales::rescale(c(-2, 0, 7.5)), 
    name = "LogFC"
  ) +
  scale_size_continuous(range = c(2, 8), name = "-log10(p-value)") +
  labs(
    x = "Gene",
    y = "Cluster A12",  
    title = "Dot plot of gene expression and significance for cluster A12",
    subtitle = "Size indicates -log10(p-value); Color indicates logFC"
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, family = "Times New Roman"), 
    text = element_text(family = "Times New Roman"), 
    legend.position = "right"
  )
```

# Bone Marrow Wilcox Test

```{r}
Idents(Seurat_Object_BM_A12) <- Seurat_Object_BM_A12$simplified_clusters
all_markers_BM <- FindAllMarkers(object = Seurat_Object_BM_A12, only.pos = T, min.pct = 0.25, min.diff.pct = 0.1, thresh.use = 0.25)
```
We print the two most relevant per cluster:
```{r}
five_top_markers_BM <- all_markers_BM %>%
  group_by(cluster) %>%
  dplyr::filter(avg_log2FC > 1 & (pct.1 > 0.25 | pct.2 > 0.25) & p_val_adj < 1e-20) %>%
  slice_min(n = 20, order_by = p_val_adj, with_ties = FALSE) %>% 
  ungroup()

five_top_markers_BM

average_expression <- AverageExpression(Seurat_Object_BM_A12, assays = "RNA", features = five_top_markers_BM$gene, return.seurat = TRUE)

DoHeatmap(
  average_expression,
  features = five_top_markers_BM$gene,
  size = 3,
  draw.lines = FALSE,
  group.colors = c(
    "Immature B" = "tan",
    "Mature B" = "tan2",
    "A12 unique" = "#E694C1",
    "Small PreB" = "lightblue3",
    "ProB" = "#A52A2A",
    "Pre-Pro B" = "palevioletred4",
    "Large PreB" = "#A694C1"
  )
) +
  scale_fill_gradientn(colors = c("grey", "red")) +
  theme(
    axis.text.y = element_text(size = 5), 
    plot.title = element_text(size = 12, family = "Times New Roman"), 
    plot.subtitle = element_text(size = 8, family = "Times New Roman"), 
    axis.title = element_text(size = 12, family = "Times New Roman"), 
    axis.text = element_text(size = 11, family = "Times New Roman") 
  )
```

# CITE-seq in Bone Marrow

```{r}
cells_to_highlight <- WhichCells(Seurat_Object_BM_selected_Bcells_idents, 
                                 expression = huIgk %in% c("huk_WT", "huk_A12"))

DimPlot(Seurat_Object_BM_selected_Bcells_idents, 
        group.by = "seurat_clusters", 
        cells.highlight = cells_to_highlight, split.by = "genotype")

positive_cells <- WhichCells(Seurat_Object_BM_A12, 
                             expression = huIgk %in% c("huk_WT", "huk_A12"))

Seurat_Object_BM_A12$huIgk_positive <- ifelse(
  Cells(Seurat_Object_BM_A12) %in% positive_cells, 
  "Positive", 
  "Negative"
)


meta_data <- Seurat_Object_BM_A12@meta.data

percentages <- meta_data %>%
  group_by(seurat_clusters_new, huIgk_positive) %>%
  summarise(n = n()) %>%
  mutate(percentage = n / sum(n) * 100)

print(percentages)

ggplot(percentages, aes(x = seurat_clusters_new, y = percentage, fill = huIgk_positive)) +
  geom_bar(stat = "identity", position = "stack") +  # Gráfico de barras apiladas
  labs(title = "Frequence of cells per cluster",
       x = "Cluster",
       y = "Pertentage",
       fill = "huIgk") +
  theme_minimal()
```

# Deeper investigation and comparison of A12 unique cluster versus small PreB and ProB

```{r}
Seurat_FrE_A12 <- subset(Seurat_Object_BM_A12, seurat_clusters_new %in% c("Small PreB", "A12 unique", "ProB"))
Idents(Seurat_FrE_A12) <- Seurat_FrE_A12$simplified_clusters
all_markers_A12_uniq_pop_BM <- FindMarkers(object = Seurat_FrE_A12, test.use = "roc", ident.1 = "A12 unique", min.pct = 0.25, min.diff.pct = 0.1, thresh.use = 0.25)

```
We print the two most relevant per cluster:
```{r}
five_top_markers_A12_uniq_pop_BM <- all_markers_A12_uniq_pop_BM %>%
                        dplyr::filter((avg_log2FC > 1 | avg_log2FC < 1) & (pct.1 > 0.25 | pct.2 > 0.25)) %>%
                        slice_max(n = 20, order_by = avg_diff) %>%  
                        slice_head(n = 20) %>% 
                        ungroup()
Idents(Seurat_FrE_A12) <- factor(
  Seurat_FrE_A12$seurat_clusters_new,
  levels = c("ProB", "Large PreB", "A12 unique", "Small PreB")
)

five_top_markers_A12_uniq_pop_BM
DoHeatmap(Seurat_FrE_A12, features =rownames(five_top_markers_A12_uniq_pop_BM)) +
  scale_fill_gradientn(colors = c("grey", "red"))

```

```{r}
ggplot(five_top_markers_A12_uniq_pop_BM, aes(x = avg_diff, y = power, label = row.names(five_top_markers_A12_uniq_pop_BM))) +
  geom_point(aes(color = myAUC), alpha = 0.7, size = 4) +
  geom_text_repel() + 
  scale_y_continuous(limits = c(0, 1), name = "Power") +
  scale_color_viridis_c(option = "D", name = "AUC", limits = c(0, 1)) + 
  labs(
    x = "Average Difference (avg_diff)"
  ) +
  theme_minimal() +
  theme(
    text = element_text(family = "Times New Roman"),
    plot.title = element_text(size = 16, face = "bold", hjust = 0.5),
    axis.text = element_text(size = 12),
    axis.title = element_text(size = 14)
  )
```

```{r}
all_markers_A12_uniq_pop_BM <- FindMarkers(object = Seurat_FrE_A12, test.use = "MAST", ident.1 = "A12 unique", min.pct = 0.25, min.diff.pct = 0.1, thresh.use = 0.25)
```

```{r}
five_top_markers_A12_uniq_pop_BM <- all_markers_A12_uniq_pop_BM %>%
                        dplyr::filter((avg_log2FC > 1 | avg_log2FC < 1) & (pct.1 > 0.25 | pct.2 > 0.25) & p_val_adj < 1e-10) %>%
                        slice_min(n = 30, order_by = p_val_adj) %>%  
                        slice_head(n = 30) %>% 
                        ungroup()
Idents(Seurat_FrE_A12) <- factor(
  Seurat_FrE_A12$seurat_clusters_new,
  levels = c("ProB", "Large PreB", "A12 unique", "Small PreB")
)

five_top_markers_A12_uniq_pop_BM
DoHeatmap(Seurat_FrE_A12, features =rownames(five_top_markers_A12_uniq_pop_BM)) +
  scale_fill_gradientn(colors = c("grey", "red"))

average_expression <- AverageExpression(Seurat_FrE_A12, features = row.names(five_top_markers_A12_uniq_pop_BM), return.seurat = TRUE)

genes_of_interest <- row.names(average_expression@assays$RNA$scale.data)

expr_subset <- average_expression@assays$RNA$scale.data[genes_of_interest, ]

ordered_genes <- rownames(expr_subset[order(expr_subset[, "A12 unique"], decreasing = TRUE), ])

DoHeatmap(average_expression, features = ordered_genes, size = 3, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red"))

DoHeatmap(average_expression, features = ordered_genes, group.colors = c(
    "ProB" = "#A52A2A",
    "Small PreB" = "lightblue3",
    "A12 unique" = "#E694C1",
    "Large PreB" = "#A694C1"
  ), size = 5, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red")) +
  theme(
    text = element_text(family = "Times New Roman"), 
    axis.text.x = element_text(size = 10, family = "Times New Roman"), 
    axis.text.y = element_text(size = 15, family = "Times New Roman"), 
    axis.title = element_text(size = 12, family = "Times New Roman"),  
    plot.title = element_text(size = 14, face = "bold", family = "Times New Roman") 
  )

average_expression <- AverageExpression(Seurat_FrE_A12, features = row.names(five_top_markers_A12_uniq_pop_BM))$RNA %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(cols = -gene, names_to = "cluster", values_to = "expression")

five_top_markers_A12_uniq_pop_BM <- five_top_markers_A12_uniq_pop_BM %>%
  dplyr::mutate(gene = rownames(five_top_markers_A12_uniq_pop_BM))

data_for_dotplot <- five_top_markers_A12_uniq_pop_BM %>%
  dplyr::select(gene, avg_log2FC, p_val_adj) %>%
  left_join(average_expression, by = "gene") %>%
  mutate(
    p_val_adj = ifelse(p_val_adj == 0, 1e-300, p_val_adj),
    significance = -log10(p_val_adj) 
  )

data_for_dotplot <- data_for_dotplot %>%
  mutate(expression_scaled = log1p(expression))  # log1p(x) = log(1 + x)

data_for_dotplot_A12 <- data_for_dotplot %>%
  filter(cluster == "A12 unique")


ggplot(data_for_dotplot_A12, aes(x = gene, y = cluster)) + 
  geom_point(aes(size = significance, color = avg_log2FC)) +  
  scale_color_gradientn(
    colors = c("blue", "white", "red"),  
    values = scales::rescale(c(-3, 0, 1.5)), 
    name = "LogFC"
  ) +
  scale_size_continuous(range = c(2, 8), name = "-log10(p-value)") +
  labs(
    x = "Gene",
    y = "Cluster A12",  # Etiqueta del eje Y solo para A12
    title = "Dot plot of gene expression and significance for cluster A12",
    subtitle = "Size indicates -log10(p-value); Color indicates logFC"
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, family = "Times New Roman"), 
    text = element_text(family = "Times New Roman"),  
    legend.position = "right"
  )
```

# A12 expresion Violin

```{r}
VlnPlot(Seurat_Object_BM_selected_Bcells_idents, features = "Igkc", group.by = "seurat_clusters_new", split.by = "genotype")
VlnPlot(Seurat_Object_SP_A12, features = "A12", group.by = "seurat_clusters_new_fin")
FeaturePlot(Seurat_Object_BM_A12, features = "Igkc")

```

# A12 expresion Barplot

```{r}
data <- Seurat_Object_BM_selected_Bcells_idents@meta.data
desired_order <- c("Pre-Pro B", "Cycling Pre-ProB", "ProB", "Cycling_ProB", 
             "Large PreB", "A12 unique", "Small PreB", "Immature B", 
             "Cycling Immature", "Mature B", "High Mitochondrial","NA")

exprs <- Seurat_Object_BM_selected_Bcells_idents@assays$RNA$counts["A12",]
data$seurat_clusters <- factor(data$seurat_clusters_new, levels = desired_order)
data$Gene_A12_Expression <- exprs > 0

prop_data <- data %>%
  group_by(seurat_clusters_new) %>%
  summarise(Proportion = mean(Gene_A12_Expression))

ggplot(prop_data, aes(x = seurat_clusters_new, y = Proportion, fill = seurat_clusters_new)) +
  geom_bar(stat = "identity") +
  labs(title = "A12 expressing cell proportion",
       x = "Cluster",
       y = "Cell proportion") +  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

# Bone Marrow Statistics

```{r}
# Obtain percentages for each cluster in each mouse
data_summary <- Seurat_Object_BM_selected_Bcells_idents@meta.data %>%
  group_by(genotype, mice, simplified_clusters) %>%
  summarize(count = n()) %>%
  mutate(percentage = count / sum(count) * 100)

# Filter out unwanted clusters
data_summary <- data_summary %>%
  filter(!is.na(simplified_clusters) & simplified_clusters != "Highly Mitochondrial")

# Create empty lists to store percentages for each genotype
A12_test <- list()
WT_test <- list()
for (i in unique(data_summary$simplified_clusters)) {
  A12_test[[as.character(i)]] <- data_summary[data_summary$genotype == "A12" & data_summary$simplified_clusters == i, ]
  WT_test[[as.character(i)]] <- data_summary[data_summary$genotype == "WT" & data_summary$simplified_clusters == i, ]
}

# Create an empty list to store t-test results
t_test_results <- list()

# Perform t-tests for each cluster between genotypes
for (i in names(A12_test)) {
  group1 <- A12_test[[i]]$percentage
  group2 <- WT_test[[i]]$percentage
  
  # Perform the t-test
  result <- t.test(group1, group2)
  
  # Store the results in the list
  t_test_results[[paste("Comparison", i)]] <- result
}

# Rename the result names to simplify
names(t_test_results) <- sub("Comparison ", "", names(t_test_results))

# Extract p-values from the t-test results
p_values <- sapply(t_test_results, function(res) res$p.value)
p_values

# Calculate log fold change and add significance markers
mean_per_cluster_A12 <- sapply(A12_test, function(df) mean(df$percentage, na.rm = TRUE))
mean_per_cluster_WT <- sapply(WT_test, function(df) mean(df$percentage, na.rm = TRUE))

# Extract p-values
p_value_list <- unlist(lapply(t_test_results, function(res) res$p.value))

# Calculate Log Fold Change (LFC)
log_fold_change <- log2(mean_per_cluster_A12 / mean_per_cluster_WT)

# Create a data frame with the results
results_df <- data.frame(
  cluster = names(A12_test),
  mean_A12 = mean_per_cluster_A12,
  mean_WT = mean_per_cluster_WT,
  log_fold_change = log_fold_change,
  p_value = p_value_list
)

# Determine levels of significance or include the p-value if between 0.05 and 0.1
results_df$significance <- ifelse(
  results_df$p_value <= 0.001, "***",
  ifelse(
    results_df$p_value <= 0.01, "**",
    ifelse(
      results_df$p_value <= 0.05, "*",
      ifelse(
        results_df$p_value <= 0.1,
        sprintf("%.3f", results_df$p_value), # Format to display p-value with 3 decimals
        ""
      )
    )
  )
)

print(results_df)

# Add significance levels to the summary data for use in the plot
data_summary <- data_summary %>%
  left_join(results_df %>% select(cluster, significance), by = c("simplified_clusters" = "cluster"))

# Define the order of clusters for the plot
cluster_order <- c("Pre-Pro B", "Cycling Pre-ProB", "ProB", "Cycling ProB", 
                   "Large PreB", "A12 unique", "Small PreB", "Immature B", 
                   "Cycling Immature", "Mature B")

data_summary$simplified_clusters <- factor(data_summary$simplified_clusters, levels = cluster_order)

# Create the plot with log fold change and significance (asterisks or p-values)
ggplot(data_summary, aes(x = simplified_clusters, y = percentage, color = genotype)) +
  geom_point(size = 3, position = position_dodge(width = 0.5), alpha = 0.7) + 
  geom_text(
    data = results_df,
    aes(x = cluster, y = max(data_summary$percentage) + 5, label = significance),
    inherit.aes = FALSE,
    size = 4,
    family = "Times New Roman"
  ) +
  scale_color_manual(values = c("WT" = "black", "A12" = "violet")) +
  labs(
    x = "Cluster",
    y = "Percentage",
    color = "Genotype"
  ) +
  theme_minimal() +
  theme(
    text = element_text(family = "Times New Roman"),
    axis.text.x = element_text(size = 11, angle = 45, hjust = 1, family = "Times New Roman"),
    legend.position = "bottom"
  )

```

# Saving work

```{r}
#saveRDS(Seurat_Object_BM_selected_Bcells_idents, file = "../Results/BM_final_Seurat_TFM_simp.RDS")
#saveRDS(Seurat_Object_SP_selected_Bcells_idents, file = "../Results/SP_final_Seurat_TFM_simp.RDS")
#Seurat_Object_BM_selected_Bcells_idents <- LoadSeuratRds("../Results/BM_final_Seurat_TFM_simp.RDS")
#Seurat_Object_SP_selected_Bcells_idents <- LoadSeuratRds("../Results/SP_final_Seurat_TFM_simp.RDS")
```

# Explore Mature clusters in bone marrow, focusing in cluster 7 related with autoimmunity

```{r}
Seurat_Mature_A12 <- subset(Seurat_Object_BM_A12, seurat_clusters %in% c("7", "0", "15", "16"))
Idents(Seurat_Mature_A12) <- Seurat_Mature_A12$seurat_clusters
all_markers_A12_uniq_pop_BM <- FindMarkers(object = Seurat_Mature_A12, test.use = "roc", ident.1 = "7", min.pct = 0.25, min.diff.pct = 0.1, thresh.use = 0.25)
```
We print the two most relevant per cluster:
```{r}
five_top_markers_A12_uniq_pop_BM <- all_markers_A12_uniq_pop_BM %>%
                        dplyr::filter((avg_log2FC > 1 | avg_log2FC < 1) & (pct.1 > 0.25 | pct.2 > 0.25)) %>%
                        slice_max(n = 20, order_by = power) %>% 
                        slice_head(n = 20) %>% 
                        ungroup()

five_top_markers_A12_uniq_pop_BM
DoHeatmap(Seurat_Mature_A12, features =rownames(five_top_markers_A12_uniq_pop_BM)) +
  scale_fill_gradientn(colors = c("grey", "red"))

average_expression <- AverageExpression(Seurat_Mature_A12, features = row.names(five_top_markers_A12_uniq_pop_BM), return.seurat = TRUE)

genes_of_interest <- row.names(average_expression@assays$RNA$scale.data)

expr_subset <- average_expression@assays$RNA$scale.data[genes_of_interest, ]

ordered_genes <- rownames(expr_subset[order(expr_subset[, "g7"], decreasing = TRUE), ])

DoHeatmap(average_expression, features = ordered_genes, size = 3, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red"))
```

```{r}
ggplot(five_top_markers_A12_uniq_pop_BM, aes(x = avg_diff, y = power, label = row.names(five_top_markers_A12_uniq_pop_BM))) +
  geom_point(aes(color = myAUC), alpha = 0.7, size = 4) + 
  geom_text_repel() + # Etiquetas de genes
  scale_y_continuous(limits = c(0, 1), name = "Power") +
  scale_color_viridis_c(option = "D", name = "AUC", limits = c(0, 1)) + 
  labs(
    x = "Average Difference (avg_diff)"
  ) +
  theme_minimal() +
  theme(
    text = element_text(family = "Times New Roman"),
    plot.title = element_text(size = 16, face = "bold", hjust = 0.5),
    axis.text = element_text(size = 12),
    axis.title = element_text(size = 14)
  )
```


```{r}
all_markers_A12_uniq_pop_BM <- FindMarkers(object = Seurat_Mature_A12, test.use = "MAST", ident.1 = "7", min.pct = 0.25, min.diff.pct = 0.1, thresh.use = 0.25)
```

```{r}
five_top_markers_A12_uniq_pop_BM <- all_markers_A12_uniq_pop_BM %>%
                        dplyr::filter((avg_log2FC > 1 | avg_log2FC < 1) & (pct.1 > 0.25 | pct.2 > 0.25) & p_val_adj < 1e-10) %>%
                        slice_min(n = 30, order_by = p_val_adj) %>%  
                        slice_head(n = 30) %>% 
                        ungroup()

five_top_markers_A12_uniq_pop_BM
DoHeatmap(Seurat_Mature_A12, features =rownames(five_top_markers_A12_uniq_pop_BM), group.by = "seurat_clusters") +
  scale_fill_gradientn(colors = c("grey", "red"))

average_expression <- AverageExpression(Seurat_Mature_A12, features = row.names(five_top_markers_A12_uniq_pop_BM), return.seurat = TRUE, group.by = "seurat_clusters")

genes_of_interest <- row.names(average_expression@assays$RNA$scale.data)

expr_subset <- average_expression@assays$RNA$scale.data[genes_of_interest, ]


DoHeatmap(average_expression, features = row.names(five_top_markers_A12_uniq_pop_BM), size = 3, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red"))

DoHeatmap(average_expression, features = row.names(five_top_markers_A12_uniq_pop_BM), group.colors = c(
    "ProB" = "#A52A2A",
    "Small PreB" = "lightblue3",
    "A12 unique" = "#E694C1",
    "Large PreB" = "#A694C1"
  ), size = 5, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red")) +
  theme(
    text = element_text(family = "Times New Roman"),  
    axis.text.x = element_text(size = 10, family = "Times New Roman"), 
    axis.text.y = element_text(size = 15, family = "Times New Roman"), 
    axis.title = element_text(size = 12, family = "Times New Roman"), 
    plot.title = element_text(size = 14, face = "bold", family = "Times New Roman")  
  )

average_expression <- AverageExpression(Seurat_Mature_A12, features = row.names(five_top_markers_A12_uniq_pop_BM), group.by = "seurat_clusters")$RNA %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(cols = -gene, names_to = "cluster", values_to = "expression")

five_top_markers_A12_uniq_pop_BM <- five_top_markers_A12_uniq_pop_BM %>%
  dplyr::mutate(gene = rownames(five_top_markers_A12_uniq_pop_BM))

data_for_dotplot_A12 <- data_for_dotplot %>%
  filter(cluster == "g7") %>%
  arrange(p_val_adj) %>%
  mutate(gene = factor(gene, levels = unique(gene))) 


ggplot(data_for_dotplot_A12, aes(x = gene, y = cluster)) +  
  geom_point(aes(size = significance, color = avg_log2FC)) + 
  scale_color_gradientn(
    colors = c("blue", "white", "red"),
    values = scales::rescale(c(-3, 0, 7)), 
    name = "LogFC"
  ) +
  scale_size_continuous(range = c(2, 8), name = "-log10(p-value)") +
  labs(
    x = "Gene",
    y = "Cluster A12", 
    title = "Dot plot of gene expression and significance for cluster A12",
    subtitle = "Size indicates -log10(p-value); Color indicates logFC"
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, family = "Times New Roman", size =12),
    text = element_text(family = "Times New Roman"),
    legend.position = "right"
  )
```

```{r}
Seurat_7_Mature <- subset(Seurat_Object_BM_selected_Bcells_idents, seurat_clusters %in% c("7"))
Idents(Seurat_7_Mature) <- Seurat_7_Mature$genotype
all_markers_A12_uniq_pop_BM <- FindMarkers(object = Seurat_7_Mature, test.use = "roc", ident.1 = "A12", min.pct = 0.25, min.diff.pct = 0.1, thresh.use = 0.25)
```
We print the two most relevant per cluster:
```{r}
five_top_markers_A12_uniq_pop_BM <- all_markers_A12_uniq_pop_BM %>%
                        dplyr::filter((avg_log2FC > 1 | avg_log2FC < 1) & (pct.1 > 0.25 | pct.2 > 0.25)) %>%
                        slice_max(n = 20, order_by = power) %>%  # Seleccionar los 15 con mayor power
                        slice_head(n = 20) %>% 
                        ungroup()

five_top_markers_A12_uniq_pop_BM
DoHeatmap(Seurat_7_Mature, features =rownames(five_top_markers_A12_uniq_pop_BM)) +
  scale_fill_gradientn(colors = c("grey", "red"))

average_expression <- AverageExpression(Seurat_7_Mature, features = row.names(five_top_markers_A12_uniq_pop_BM), return.seurat = TRUE)

genes_of_interest <- row.names(average_expression@assays$RNA$scale.data)

expr_subset <- average_expression@assays$RNA$scale.data[genes_of_interest, ]

DoHeatmap(average_expression, features = row.names(five_top_markers_A12_uniq_pop_BM), size = 3, draw.lines = FALSE) +
  scale_fill_gradientn(colors = c("grey", "red"))
```

```{r}
ggplot(five_top_markers_A12_uniq_pop_BM, aes(x = avg_diff, y = power, label = row.names(five_top_markers_A12_uniq_pop_BM))) +
  geom_point(aes(color = myAUC), alpha = 0.7, size = 4) + # Aumentar el tamaño de los puntos
  geom_text_repel() + # Etiquetas de genes
  scale_y_continuous(limits = c(0, 1), name = "Power") + # Escala para Power de 0 a 1
  scale_color_viridis_c(option = "D", name = "AUC", limits = c(0, 1)) + # Escala de color para AUC de 0 a 1
  labs(
    x = "Average Difference (avg_diff)"
  ) +
  theme_minimal() +
  theme(
    text = element_text(family = "Times New Roman"),
    plot.title = element_text(size = 16, face = "bold", hjust = 0.5),
    axis.text = element_text(size = 12),
    axis.title = element_text(size = 14)
  )
```


```{r}
all_markers_A12_uniq_pop_BM <- FindMarkers(object = Seurat_7_Mature, test.use = "MAST", ident.1 = "A12", min.pct = 0.25, min.diff.pct = 0.1, thresh.use = 0.25)
```

```{r}
five_top_markers_A12_uniq_pop_BM <- all_markers_A12_uniq_pop_BM %>%
                        dplyr::filter((avg_log2FC > 1 | avg_log2FC < 1) & (pct.1 > 0.25 | pct.2 > 0.25) & p_val_adj < 1e-10) %>%
                        slice_min(n = 30, order_by = p_val_adj) %>%  # Seleccionar los 15 con menor p_val_adj
                        slice_head(n = 30) %>% 
                        ungroup()

five_top_markers_A12_uniq_pop_BM
DoHeatmap(Seurat_7_Mature, features =rownames(five_top_markers_A12_uniq_pop_BM), group.by = "seurat_clusters") +
  scale_fill_gradientn(colors = c("grey", "red"))

```

# Proportions bone marrow per each cluster (numeric cluster designated by umap) grouped by cell subtype clusters

```{r}
# Obtain percentages for each cluster in each mice
data_summary <- Seurat_Object_BM_selected_Bcells_idents@meta.data %>%
  group_by(genotype, mice, seurat_clusters) %>%
  summarize(count = n()) %>%
  mutate(percentage = count / sum(count) * 100)

# Store percentages in lists for each genotype

data_summary <- data_summary %>%
  filter(!is.na(seurat_clusters) & seurat_clusters != "17")
A12_test <- list()
WT_test <- list()
for (i in unique(data_summary$seurat_clusters)) {
  A12_test[[as.character(i)]] <- data_summary[data_summary$genotype == "A12" & data_summary$seurat_clusters == i, ]
  WT_test[[as.character(i)]] <- data_summary[data_summary$genotype == "WT" & data_summary$seurat_clusters == i, ]
}

# Crear una lista vacía para almacenar los resultados
resultados_t_test <- list()

# Iterar sobre las listas y realizar la prueba t para cada par de elementos
for (i in names(A12_test)) {
  grupo1 <- A12_test[[i]]$percentage
  grupo2 <- WT_test[[i]]$percentage
  
  # Realizar la prueba t
  resultado <- t.test(grupo1, grupo2, var.equal = TRUE)
  
  # Almacenar los resultados en la lista
  resultados_t_test[[paste("Comparacion", i)]] <- resultado
}

# Imprimir los resultados
names(resultados_t_test) <- sub("Comparacion ", "", names(resultados_t_test))

p_values <- sapply(resultados_t_test, function(res) res$p.value)
p_values

# Calcular log fold change y añadir asteriscos de significancia
mean_per_cluster_A12 <- sapply(A12_test, function(df) mean(df$percentage, na.rm = TRUE))
mean_per_cluster_WT <- sapply(WT_test, function(df) mean(df$percentage, na.rm = TRUE))

p_value_l <- unlist(lapply(resultados_t_test, function(res) res$p.value))

# Calcular Log Fold Change (LFC)
log_fold_change <- log2(mean_per_cluster_A12 / mean_per_cluster_WT)

# Crear un data frame con todos los resultados
resultados_df <- data.frame(
  cluster = names(A12_test),
  mean_A12 = mean_per_cluster_A12,
  mean_WT = mean_per_cluster_WT,
  log_fold_change = log_fold_change,
  p_value = p_value_l
)

# Determinar niveles de significancia o incluir el valor de p si está entre 0.05 y 0.1
resultados_df$significance <- ifelse(
  resultados_df$p_value <= 0.001, "***",
  ifelse(
    resultados_df$p_value <= 0.01, "**",
    ifelse(
      resultados_df$p_value <= 0.05, "*",
      ifelse(
        resultados_df$p_value <= 0.1,
        sprintf("%.3f", resultados_df$p_value), # Formato para mostrar p-value con 3 decimales
        ""
      )
    )
  )
)

print(resultados_df)

# Agregar los niveles de significancia al resumen para usarlos en el gráfico
data_summary <- data_summary %>%
  left_join(resultados_df %>% select(cluster, significance), by = c("seurat_clusters" = "cluster"))



data_summary$seurat_clusters <- factor(data_summary$seurat_clusters)
# Paso 1: Asignar los grupos de clusters a data_summary
data_summary$cluster_group <- case_when(
  data_summary$seurat_clusters %in% c(5, 9) ~ "Pre-Pro B",  
  data_summary$seurat_clusters %in% c(12, 11) ~ "ProB",
  data_summary$seurat_clusters %in% c(8, 6, 10, 14) ~ "Large PreB",
  data_summary$seurat_clusters %in% c(2, 3) ~ "Small PreB",
  data_summary$seurat_clusters == 4 ~ "A12 unique",
  data_summary$seurat_clusters == 1 ~ "Immature B",
  data_summary$seurat_clusters %in% c(0, 7, 15, 16) ~ "Mature B",
  TRUE ~ "Excluded"  # Para el cluster 17 que se excluye
)

# Paso 2: Asignar colores a cada grupo con los colores específicos
cluster_colors <- c(
  "Pre-Pro B" = "palevioletred4",
  "ProB" = "#A52A2A",
  "Large PreB" = "#A694C1",
  "A12 unique" = "#E694C1",
  "Small PreB" = "lightblue3",
  "Immature B" = "tan",
  "Mature B" = "tan2"
)

# Paso 3: Convertir 'cluster_group' a un factor con el orden deseado
data_summary$cluster_group <- factor(data_summary$cluster_group, 
                                     levels = c(
                                       "Pre-Pro B", "ProB", "Large PreB", 
                                       "Small PreB", "A12 unique", "Immature B", 
                                       "Mature B"
                                     ))

# Convertir 'seurat_clusters' a factor para asegurar el orden de los clusters según el 'cluster_group'
data_summary$seurat_clusters <- factor(data_summary$seurat_clusters, 
                                       levels = unique(data_summary$seurat_clusters[order(data_summary$cluster_group)]))

# Paso 4: Crear el gráfico
ggplot(data_summary, aes(x = seurat_clusters, y = percentage, color = genotype)) +
  geom_point(size = 3, position = position_dodge(width = 0.5), alpha = 0.7) + 
  geom_text(
    data = resultados_df,
    aes(x = cluster, y = max(data_summary$percentage) + 5, label = significance),
    inherit.aes = FALSE,
    size = 4,
    family = "Times New Roman"
  ) +
  scale_color_manual(values = c("WT" = "black", "A12" = "violet")) +  # Colorear puntos por genotipo
  scale_x_discrete(
    breaks = levels(data_summary$seurat_clusters), 
    labels = function(x) {
      # Asignar colores a las etiquetas del eje X según el grupo del cluster
      sapply(x, function(cluster) {
        group <- data_summary$cluster_group[data_summary$seurat_clusters == cluster][1]
        color <- cluster_colors[group]
        # Crear un texto con color y negrita (usando 'ggtext' para colores)
        return(paste0("<span style='color:", color, "; font-weight: bold;'>", cluster, "</span>"))
      })
    }
  ) + 
  labs(
    x = "Cluster",
    y = "Percentage",
    color = "Genotype"
  ) +
  theme_minimal() +
  theme(
    text = element_text(family = "Times New Roman"),
    axis.text.x = element_markdown(size = 14, angle = 45, hjust = 1, family = "Times New Roman"),  # Aumentamos el tamaño de la fuente y la negrita
    legend.position = "bottom"
  )
```

# Anergy signature

```{r}
# Genes obtained from anergic population RNAseq: # Multiple tolerance checkpoints restrain affinity maturation of B cells expressing the germlineprecursor of a lupus patient-derived anti-dsDNA antibody in knock-in mice

Anergic_marker_gene_list <- list(c("Fgl2", "Rgl1", "Gcsam", "Slc9c1", "Serpina3f", "Snrnp25", "Lilrb4a", 
           "Thsd7a", "Egr2", "Lrmp", "Serpinb9", "Marcksl1", "Il9r", "Gm31763", 
           "Socs1", "Cd72", "Serpina3g", "Ndrg1", "Lag3", "Ccdc17", "Anxa3", 
           "Aox4", "Fut4", "Fbxl21", "Zbp1", "Apoe", "Ctse", "Wdfy1", "Cxcr4", 
           "2210039B01Rik", "Gca", "Ptprv", "Chst7", "Tmem106b", "Fam3c", "Ephx1", 
           "Dnm3", "Serpinb6b", "Gins2", "Ptchd1", "H2-Oa", "Syt11", "Chst10", 
           "Gbp7", "Fam129a", "H2-Q7", "Loxhd1", "Ackr3", "Traf1", "Ffar2", "Hivep3", "Rhbdf1", "Alcam", 
             "Dusp4", "Btn2a2", "Mafb", "Camp", "Pard3b", "Nlrc5", "Slc2a6", 
             "Chka", "2810417H13Rik", "Litaf", "Twsg1", "Lck", "Amigo2", "Tgfb3", 
             "Hcrtr2", "Phf11a", "Glt28d2", "Jchain", "Ms4a6c", "Nsg1", "Iigp1", 
             "Plin2", "Vav3", "Tmem151b", "Faah", "Snhg18", "Ccl4", "Cdca7l", 
             "Phf11b", "Vcam1", "Fam160a1", "Parm1", "Bend6", "Dtx3", "Batf", 
             "Il12a", "Fscn1", "Mpo", "Dyrk4", "Hp", "Slco5a1", "Zfp30", "Daam1")) 

Seurat_Object_SP_A12 <- AddModuleScore(
  object = Seurat_Object_SP_A12, 
  features = Anergic_marker_gene_list, 
  name = "Anergic_score"
)
Seurat_Object_SP_WT <- AddModuleScore(
  object = Seurat_Object_SP_WT, 
  features = Anergic_marker_gene_list, 
  name = "Anergic_score"
)

plot_A12 <- FeaturePlot(Seurat_Object_SP_WT, features = "Anergic_score1", split.by = "genotype") + 
  scale_color_viridis(option = "D") + 
  ggtitle("AnergicB Score", subtitle = "Genes: Flt3, Tyrobp, Irf8, Il7r") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16),  
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )

plot_WT <- FeaturePlot(Seurat_Object_SP_A12, features = "Anergic_score1", split.by = "genotype") + 
  scale_color_viridis(option = "D") + 
  ggtitle("AnergicB Score", subtitle = "Genes: Flt3, Tyrobp, Irf8, Il7r") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16),  
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )

wrap_plots(plot_A12, plot_WT, ncol = 2)  

#Calculate mean anergic score

mean_scores_A12 <- Seurat_Object_SP_A12@meta.data %>%
  group_by(seurat_clusters_new_fin, mice) %>%
  summarise(mean_Anergic_score1 = mean(Anergic_score1, na.rm = TRUE))

mean_scores_A12

# Apply ANOVA for the variable Anergic_score1 for seurat_clusters_new_fin
anova_result_A12 <- aov(mean_Anergic_score1 ~ seurat_clusters_new_fin, data = mean_scores_A12)

# Obtain summary of ANOVA
anova_summary_A12 <- summary(anova_result_A12)

# Extract p values
anova_summary_A12[[1]]

# Apply test of Tukey post-hoc
tukey_result_A12 <- TukeyHSD(anova_result_A12)

plot(tukey_result_A12)

# Pairwise t-test
t_test_result_A12 <- pairwise.t.test(
  mean_scores_A12$mean_Anergic_score1, 
  mean_scores_A12$seurat_clusters_new_fin,
  p.adjust.method = "BH"  
)

t_test_pvals <- t_test_result_A12$p.value

significant_comparisons <- data.frame(
  group1 = rep(rownames(t_test_pvals), ncol(t_test_pvals)),
  group2 = rep(colnames(t_test_pvals), each = nrow(t_test_pvals)),
  p_value = as.vector(t_test_pvals)
)

significant_comparisons <- significant_comparisons %>%
  filter(p_value < 1e-6)

comparisons <- split(significant_comparisons, seq(nrow(significant_comparisons))) %>%
  lapply(function(x) c(x$group1, x$group2))

assign_asterisks <- function(p) {
  if (p < 0.001) {
    return("***")
  } else if (p < 0.01) {
    return("**")
  } else if (p < 0.05) {
    return("*")
  } else {
    return("ns")  # No significativo
  }
}

annotations <- sapply(significant_comparisons$p_value, assign_asterisks)

y_positions <- seq(from = max(mean_scores_A12$mean_Anergic_score1) + 0.05, 
                   by = 0.01, length.out = length(comparisons))

cluster_colors <- c(
  "Fo B" = "#D55E00",
  "Transitional B" = "darkred",
  "MZ B" = "#009E73",
  "A12 expressing" = "#DC79A7",
  "Mix" = "darkslategray",
  "GC B" = "tan4",
  "B1" = "#ED8987A7"
)

ggplot(mean_scores_A12, aes(x = seurat_clusters_new_fin, y = mean_Anergic_score1, color = seurat_clusters_new_fin)) +
  geom_point(size = 3, alpha = 0.7) +  # Puntos
  labs(x = "Cluster", y = "Anergic Score", title = "Anergic Score distribution per cluster") +
  theme_minimal() +
  scale_color_manual(values = cluster_colors) +  
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  geom_signif(
    comparisons = comparisons, 
    annotations = annotations, 
    map_signif_level = FALSE,  
    y_position = y_positions,   
    step_increase = 0.1       
  )
```

```{r}
# Anergic gene signature obtained from: IgD attenuates the IgM-induced anergy response in transitional and mature B cells

Anergic_marker_gene_list <- list(
  unique(c("Ndrg1", "Rasgef1a", "Rgs13", "Gfi1", "Egr2", "Sdc1", "Myb", "Chst7", 
           "Tacc2", "Gad1", "Hpgds", "Rgl1", "Bcl6b", "Fosb", "Ms4a4a", "NAP114472-1", 
           "Eno2", "Mreg", "Egr3", "Lonrf3", "9430078K24Rik", "Rgl1", "Endou", "Loxhd1", 
           "Nrgn", "Fabp5", "Axl", "LOC668306", "Fos", "Crisp3", "Rgs1", "Litaf", "Lck", 
           "Nr4a1", "Fos", "Egr1", "Lrfn4", "Ahr", "NAP112057-1", "Fos", "Pcp4", "Fos", 
           "Slc9b2", "Ppnr", "Sh3bgr", "Acoxl", "Fos", "Fos", "Ets2", "Npdc1", "Lass6", 
           "Rgs2", "Akap5", "Rgs2", "A_55_P1961081", "Dusp4", "Lzts2", "Dll1", "Car2", 
           "Igf1r", "Fam160a1", "Fos", "Cd63", "Ephx1", "Rapgef3", "Zfpm1", "Zfpm1", "Fos", 
           "Sox4", "Sh3bgr", "Fos", "Lef1", "Pgk2", "Lef1", "AY351698", "Sox4", "6330403A02Rik", 
           "Faah", "Dll1", "Pgpep1l", "Apon", "Derl3", "Ankrd13b", "Ephx1", "Crip3", "Cenpv", 
           "Fos", "Dll1", "Ifng", "Sox4", "Dll1", "Dll1", "Tmem151b", "Shmt1", "Nab2", "Dll1", 
           "Alcam", "St8sia6", "Epha2", "Dll1", "Serp2", "Dll1", "Echdc3", "Astl", "Zfp948", 
           "Syne1", "Echdc3", "Sh2d5", "8430427H17Rik", "Pacsin1", "Zfp52", "Sestd1", "Tyms", 
           "Zfp318", "Tacstd2", "Apon", "Dll1", "Specc1", "Zfp318", "Aqp3", "Csdc2", "Dll1", 
           "Grap2", "Apon", "Rgs3", "Slc4a11", "Rassf6", "Fam167a", "B3gnt7", "Apon", "Apoe", 
           "Renbp", "Apoe", "9630013D21Rik", "Apoe", "Apoe", "Dhrs3", "Apoe", "Sh2d5", "Klc3", 
           "Apoe", "Apoe", "Apoe", "Abcb4", "Faim3", "Nab2", "Reep2", "Apoe", "Cldn10", 
           "Cdk6", "Apoe", "Gm2447", "Klf10", "Atp10a", "Usp2", "Renbp", "Mgll", "Il4i1", 
           "Nefh", "Omp", "Pdzd2", "Cacna1d", "Marcksl1", "Ccnd2", "X79554", "Dcbld1", "Dcbld1", 
           "Plxna3", "LOC100040974", "Vmn2r60", "A_55_P2010134", "Ptchd1"))
)

Seurat_Object_SP_A12 <- AddModuleScore(
  object = Seurat_Object_SP_A12, 
  features = Anergic_marker_gene_list, 
  name = "Anergic_score"
)
Seurat_Object_SP_WT <- AddModuleScore(
  object = Seurat_Object_SP_WT, 
  features = Anergic_marker_gene_list, 
  name = "Anergic_score"
)

plot_A12 <- FeaturePlot(Seurat_Object_SP_WT, features = "Anergic_score1", split.by = "genotype") + 
  scale_color_viridis(option = "D") + 
  ggtitle("AnergicB Score", subtitle = "Genes: Flt3, Tyrobp, Irf8, Il7r") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16),  
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )

plot_WT <- FeaturePlot(Seurat_Object_SP_A12, features = "Anergic_score1", split.by = "genotype") + 
  scale_color_viridis(option = "D") + 
  ggtitle("AnergicB Score", subtitle = "Genes: Flt3, Tyrobp, Irf8, Il7r") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16),  
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )

wrap_plots(plot_A12, plot_WT, ncol = 2)  

mean_scores_A12 <- Seurat_Object_SP_A12@meta.data %>%
  group_by(seurat_clusters_new_fin, mice) %>%
  summarise(mean_Anergic_score1 = mean(Anergic_score1, na.rm = TRUE))

mean_scores_A12

anova_result_A12 <- aov(mean_Anergic_score1 ~ seurat_clusters_new_fin, data = mean_scores_A12)

anova_summary_A12 <- summary(anova_result_A12)

anova_summary_A12[[1]]

tukey_result_A12 <- TukeyHSD(anova_result_A12)
plot(tukey_result_A12)


t_test_result_A12 <- pairwise.t.test(
  mean_scores_A12$mean_Anergic_score1, 
  mean_scores_A12$seurat_clusters_new_fin,
  p.adjust.method = "BH" 
)

t_test_pvals <- t_test_result_A12$p.value

significant_comparisons <- data.frame(
  group1 = rep(rownames(t_test_pvals), ncol(t_test_pvals)),
  group2 = rep(colnames(t_test_pvals), each = nrow(t_test_pvals)),
  p_value = as.vector(t_test_pvals)
)

significant_comparisons <- significant_comparisons %>%
  filter(p_value < 1e-6)

comparisons <- split(significant_comparisons, seq(nrow(significant_comparisons))) %>%
  lapply(function(x) c(x$group1, x$group2))

assign_asterisks <- function(p) {
  if (p < 0.001) {
    return("***")
  } else if (p < 0.01) {
    return("**")
  } else if (p < 0.05) {
    return("*")
  } else {
    return("ns")  # No significativo
  }
}

annotations <- sapply(significant_comparisons$p_value, assign_asterisks)

y_positions <- seq(from = max(mean_scores_A12$mean_Anergic_score1) + 0.05, 
                   by = 0.01, length.out = length(comparisons))

cluster_colors <- c(
  "Fo B" = "#D55E00",
  "Transitional B" = "darkred",
  "MZ B" = "#009E73",
  "A12 expressing" = "#DC79A7",
  "Mix" = "darkslategray",
  "GC B" = "tan4",
  "B1" = "#ED8987A7"
)

ggplot(mean_scores_A12, aes(x = seurat_clusters_new_fin, y = mean_Anergic_score1, color = seurat_clusters_new_fin)) +
  geom_point(size = 3, alpha = 0.7) +  # Puntos
  labs(x = "Cluster", y = "Anergic Score", title = "Distribución de Anergic Score por Clúster") +
  theme_minimal() +
  scale_color_manual(values = cluster_colors) +  
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +  
  geom_signif(
    comparisons = comparisons,  
    annotations = annotations, 
    map_signif_level = FALSE,   
    y_position = y_positions, 
    step_increase = 0.1        
  )
```

# ABC marker gene list

```{r}
ABC_marker_gene_list <- c("Itgb1", "Anxa2", "S100a10", "H2-Eb1", "Ly6c2", "Zbtb32", "Vim", 
           "Sat1", "H2-Aa", "Ahnak", "Zbtb20", "Crip1", "Scimp", "Ly6a", 
           "Cd74", "Tmsb4x", "Ass1", "Dnajc7", "Fcer1g", "Pdlim1", 
           "Slc25a19", "Fxyd5", "Lgals3", "Fam3c", "Ctsh", "Rgs1", "Rcn3", 
           "Tcf4", "Tagln2", "Ly6e", "Zeb2", "Laptm5", "Spn", "Sp140", 
           "Lsp1", "Cd2")

Seurat_Object_SP_A12 <- AddModuleScore(
  object = Seurat_Object_SP_A12, 
  features = Anergic_marker_gene_list, 
  name = "Anergic_score"
)
Seurat_Object_SP_WT <- AddModuleScore(
  object = Seurat_Object_SP_WT, 
  features = Anergic_marker_gene_list, 
  name = "Anergic_score"
)

plot_A12 <- FeaturePlot(Seurat_Object_SP_WT, features = "Anergic_score1", split.by = "genotype") + 
  scale_color_viridis(option = "D") + 
  ggtitle("AnergicB Score", subtitle = "Genes: Flt3, Tyrobp, Irf8, Il7r") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16),  
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )

plot_WT <- FeaturePlot(Seurat_Object_SP_A12, features = "Anergic_score1", split.by = "genotype") + 
  scale_color_viridis(option = "D") + 
  ggtitle("AnergicB Score", subtitle = "Genes: Flt3, Tyrobp, Irf8, Il7r") +  
  theme(
    plot.title = element_text(family = "Times New Roman", size = 16),  
    plot.subtitle = element_text(family = "Times New Roman", size = 14, hjust = 0.5)  
  )

wrap_plots(plot_A12, plot_WT, ncol = 2)  

mean_scores_A12 <- Seurat_Object_SP_A12@meta.data %>%
  group_by(seurat_clusters_new_fin, mice) %>%
  summarise(mean_Anergic_score1 = mean(Anergic_score1, na.rm = TRUE))

mean_scores_A12

anova_result_A12 <- aov(mean_Anergic_score1 ~ seurat_clusters_new_fin, data = mean_scores_A12)

anova_summary_A12 <- summary(anova_result_A12)


anova_summary_A12[[1]]  
tukey_result_A12 <- TukeyHSD(anova_result_A12)

plot(tukey_result_A12)


t_test_result_A12 <- pairwise.t.test(
  mean_scores_A12$mean_Anergic_score1, 
  mean_scores_A12$seurat_clusters_new_fin,
  p.adjust.method = "BH" 
)

t_test_pvals <- t_test_result_A12$p.value

significant_comparisons <- data.frame(
  group1 = rep(rownames(t_test_pvals), ncol(t_test_pvals)),
  group2 = rep(colnames(t_test_pvals), each = nrow(t_test_pvals)),
  p_value = as.vector(t_test_pvals)
)

significant_comparisons <- significant_comparisons %>%
  filter(p_value < 0.05)

comparisons <- split(significant_comparisons, seq(nrow(significant_comparisons))) %>%
  lapply(function(x) c(x$group1, x$group2))

assign_asterisks <- function(p) {
  if (p < 0.001) {
    return("***")
  } else if (p < 0.01) {
    return("**")
  } else if (p < 0.05) {
    return("*")
  } else {
    return("ns") 
  }
}

annotations <- sapply(significant_comparisons$p_value, assign_asterisks)
y_positions <- seq(from = max(mean_scores_A12$mean_Anergic_score1) + 0.05, 
                   by = 0.01, length.out = length(comparisons))
cluster_colors <- c(
  "Fo B" = "#D55E00",
  "Transitional B" = "darkred",
  "MZ B" = "#009E73",
  "A12 expressing" = "#DC79A7",
  "Mix" = "darkslategray",
  "GC B" = "tan4",
  "B1" = "#ED8987A7"
)

ggplot(mean_scores_A12, aes(x = seurat_clusters_new_fin, y = mean_Anergic_score1, color = seurat_clusters_new_fin)) +
  geom_point(size = 3, alpha = 0.7) +  # Puntos
  labs(x = "Cluster", y = "Anergic Score", title = "Anergic score distribution per cluster") +
  theme_minimal() +
  scale_color_manual(values = cluster_colors) + 
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  geom_signif(
    comparisons = comparisons, 
    annotations = annotations, 
    map_signif_level = FALSE,  
    y_position = y_positions,  
    step_increase = 0.1        
  )
```




