---
title: "RNAvelocity_Seurat_Processing"
author: "Javier Molina Valenzuela"
date: "2025-01-08"
output: html_document
---

# Processing Seurat Object for RNA Velocity
Referring to the guide at this blog post,

The Seurat_Object_BM_selected_Bcells_idents$barcodes variable will store the names of the cells.

Why are there two UMAP dimensions (UMAP_1 and UMAP_2)?
The UMAP algorithm (Uniform Manifold Approximation and Projection) is used to reduce the dimensionality of high-dimensional data, allowing complex structures to be visualized in a low-dimensional space. Typically, UMAP is used to project data into a 2-dimensional (2D) space for visualization, although it can also project into 3D or higher dimensions, depending on the configuration.

When performing UMAP dimensionality reduction to 2 dimensions, you obtain two principal components:

UMAP_1: The first dimension of the UMAP projection.
UMAP_2: The second dimension of the UMAP projection.
These two dimensions (UMAP_1 and UMAP_2) represent the coordinates of each cell in the 2D space generated by the UMAP algorithm. The combination of UMAP_1 and UMAP_2 allows you to visualize the data in a 2D scatter plot, where:

Each point corresponds to a single cell.
The position of the points reflects similarities and differences between cells based on their gene expression profiles.
In summary, UMAP_1 and UMAP_2 provide the necessary components to visualize high-dimensional biological data in a way that is easy to interpret and analyze in two dimensions.

```{r}
Seurat_Object_BM_selected_Bcells_idents$immgen_BM.main <- gsub("B cells, pro", "B cells pro", Seurat_Object_BM_selected_Bcells_idents$immgen_BM.main)

Seurat_Object_BM_selected_Bcells_idents$immgen_BM.fine <- gsub("B cells,\\s*pro \\(proB\\.FrA\\)", "B cells pro (proB.FrA)", Seurat_Object_BM_selected_Bcells_idents$immgen_BM.fine, perl = TRUE)
DimPlot(Seurat_Object_BM_selected_Bcells_idents, reduction = "umap", group.by = "immgen_BM.fine")
# save metadata table:
Seurat_Object_BM_selected_Bcells_idents$barcodes <- colnames(Seurat_Object_BM_selected_Bcells_idents)
Seurat_Object_BM_selected_Bcells_idents$UMAP_1 <- Seurat_Object_BM_selected_Bcells_idents@reductions$umap@cell.embeddings[,1]
Seurat_Object_BM_selected_Bcells_idents$UMAP_2 <- Seurat_Object_BM_selected_Bcells_idents@reductions$umap@cell.embeddings[,2]
write.csv(Seurat_Object_BM_selected_Bcells_idents@meta.data, file='../../RNAvel/TFM_Data/metadata.csv', quote=F, row.names=F)

# write expression counts matrix
counts_matrix <- GetAssayData(Seurat_Object_BM_selected_Bcells_idents, assay='RNA', slot='counts')
writeMM(counts_matrix, file= "../../RNAvel/Data/counts.mtx")

# write dimesnionality reduction matrix, in this example case pca matrix
write.csv(Seurat_Object_BM_selected_Bcells_idents@reductions$pca@cell.embeddings, file='../../RNAvel/TFM_Data/pca_BM.csv', quote=F, row.names=F)

# write gene names
write.table(
  data.frame('gene'=rownames(counts_matrix)),file='../../RNAvel/TFM_Data/gene_names_BM.csv',
  quote=F,row.names=F,col.names=F
)
```


```{r}
Seurat_Object_BM_A12 <- subset(Seurat_Object_BM_selected_Bcells_idents, genotype == "A12")
Seurat_Object_BM_WT <- subset(Seurat_Object_BM_selected_Bcells_idents, genotype == "WT")
Seurat_Object_BM_A12$immgen_BM.main <- gsub("B cells, pro", "B cells pro", Seurat_Object_BM_A12$immgen_BM.main)

Seurat_Object_BM_A12$immgen_BM.fine <- gsub("B cells,\\s*pro \\(proB\\.FrA\\)", "B cells pro (proB.FrA)", Seurat_Object_BM_A12$immgen_BM.fine, perl = TRUE)
DimPlot(Seurat_Object_BM_A12, reduction = "umap", group.by = "immgen_BM.fine")
# save metadata table:
Seurat_Object_BM_A12$barcodes <- colnames(Seurat_Object_BM_A12)
Seurat_Object_BM_A12$UMAP_1 <- Seurat_Object_BM_A12@reductions$umap@cell.embeddings[,1]
Seurat_Object_BM_A12$UMAP_2 <- Seurat_Object_BM_A12@reductions$umap@cell.embeddings[,2]
write.csv(Seurat_Object_BM_A12@meta.data, file='../../RNAvel/TFM_Data/metadata_A12.csv', quote=F, row.names=F)

# write expression counts matrix
counts_matrix <- GetAssayData(Seurat_Object_BM_A12, assay='RNA', slot='counts')
writeMM(counts_matrix, file= "../../RNAvel/TFM_Data/counts_A12.mtx")

# write dimesnionality reduction matrix
write.csv(Seurat_Object_BM_A12@reductions$pca@cell.embeddings, file='../../RNAvel/TFM_Data/pca_A12.csv', quote=F, row.names=F)

# write gene names
write.table(
  data.frame('gene'=rownames(counts_matrix)),file='../../RNAvel/TFM_Data/gene_names_A12.csv',
  quote=F,row.names=F,col.names=F
)
```

```{r}
Seurat_Object_BM_WT$immgen_BM.main <- gsub("B cells, pro", "B cells pro", Seurat_Object_BM_WT$immgen_BM.main)

Seurat_Object_BM_WT$immgen_BM.fine <- gsub("B cells,\\s*pro \\(proB\\.FrA\\)", "B cells pro (proB.FrA)", Seurat_Object_BM_WT$immgen_BM.fine, perl = TRUE)
DimPlot(Seurat_Object_BM_WT, reduction = "umap", group.by = "immgen_BM.fine")
# save metadata table:
Seurat_Object_BM_WT$barcodes <- colnames(Seurat_Object_BM_WT)
Seurat_Object_BM_WT$UMAP_1 <- Seurat_Object_BM_WT@reductions$umap@cell.embeddings[,1]
Seurat_Object_BM_WT$UMAP_2 <- Seurat_Object_BM_WT@reductions$umap@cell.embeddings[,2]
write.csv(Seurat_Object_BM_WT@meta.data, file='../../RNAvel/TFM_Data/metadata_WT.csv', quote=F, row.names=F)

# write expression counts matrix
counts_matrix <- GetAssayData(Seurat_Object_BM_WT, assay='RNA', slot='counts')
writeMM(counts_matrix, file= "../../RNAvel/TFM_Data/counts_WT.mtx")

# write dimesnionality reduction matrix, in this example case pca matrix
write.csv(Seurat_Object_BM_WT@reductions$pca@cell.embeddings, file='../../RNAvel/TFM_Data/pca_WT.csv', quote=F, row.names=F)

# write gene names
write.table(
  data.frame('gene'=rownames(counts_matrix)),file='../../RNAvel/TFM_Data/gene_names_WT.csv',
  quote=F,row.names=F,col.names=F
)
```
